<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"  
  "http://www.w3.org/TR/html4/loose.dtd">  
<html > 
<head><title></title> 
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"> 
<meta name="generator" content="TeX4ht (http://www.cse.ohio-state.edu/~gurari/TeX4ht/)"> 
<meta name="originator" content="TeX4ht (http://www.cse.ohio-state.edu/~gurari/TeX4ht/)"> 
<!-- html --> 
<meta name="src" content="steventhesis_num.tex"> 
<meta name="date" content="2011-04-30 22:46:00"> 
<link rel="stylesheet" type="text/css" href="steventhesis_num.css"> 
</head><body 
>
<div class="center" 
>
<!--l. 133--><p class="noindent" >
<!--l. 133--><p class="noindent" >A MODEL FOR THE EVOLUTION OF NUCLEOTIDE POLYMERASE
DIRECTIONALITY<br />
by<br />
Joshua Ballanco<br />
A DISSERTATION<br />
<!--l. 133--><p class="noindent" >Submitted to the Faculty of the Stevens Institute of Technology<br />
in partial fulfillment of the requirements for the degree of<br /><br />
<!--l. 133--><p class="noindent" >DOCTOR OF PHILOSOPHY<br />

<!--l. 133--><p class="noindent" >                    <div class="minipage">Joshua Ballanco, Candidate<br 
class="newline" /> <span class="underline">ADVISORY COMMITTEE</span><br 
class="newline" /> Marc Mansfield, Chairman Date<br 
class="newline" /> Philip Leopold Date<br 
class="newline" /> A. K. Ganguly Date<br 
class="newline" /> Nicholas Murgolo Date<br 
class="newline" /> Knut Stamnes Date<br 
class="newline" />
</div>
<!--l. 133--><p class="noindent" >
<!--l. 133--><p class="noindent" >STEVENS INSTITUTE OF TECHNOLOGY<br />
Castle Point on Hudson<br />
Hoboken, NJ 07030<br />
2011
</div>
<!--l. 133--><p class="noindent" >

<!--l. 135--><p class="indent" >

<div class="center" 
>
<!--l. 135--><p class="noindent" >
<!--l. 135--><p class="noindent" ><span 
class="cmsy-10x-x-120">©</span>Copyright 2011&#x00A0;Joshua Ballanco</div>
<!--l. 135--><p class="noindent" >

<div class="center" 
>
<!--l. 137--><p class="noindent" >
<!--l. 137--><p class="noindent" >A MODEL FOR THE EVOLUTION OF NUCLEOTIDE POLYMERASE
DIRECTIONALITY<br />
ABSTRACT<br />
</div>
<!--l. 137--><p class="noindent" >In all known living organisms, every enzyme that synthesizes nucleic acid polymers does so by
adding nucleotide 5&#8217;-triphosphates to the 3&#8217;-hydroxyl group of the growing chain. This
results in the well known 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>directionality of all DNA and RNA Polymerases. The lack
of any alternative mechanism, e.g. addition in a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>direction, may indicate a very
early founder effect in the evolution of life, or it may be the result of a selective
pressure against such an alternative. In an attempt to determine whether the lack
of an alternative polymerase directionality is the result of a founder effect or
evolutionary selection, we have constructed a basic model of early polymerase evolution.
This model is informed by the essential chemical properties of the nucleotide
polymerization reaction. With this model, we are able to simulate the growth of
organisms with polymerases that synthesize either 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>or 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>in isolation
or in competition with each other. We have found that a competition between
organisms with 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerases and 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerases only results in a
evolutionarily stable strategy under certain conditions. Furthermore, we have found that
mutations lead to a much clearer delineation between conditions that lead to a stable
coexistence of these populations and conditions which ultimately lead to success for the
5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>form. In addition to presenting a plausible explanation for the uniqueness
of enzymatic polymerization reactions, we hope these results also provide an

example of how whole organism evolution can be understood based on molecular
details.

<div class="flushleft" 
>
<!--l. 137--><p class="noindent" >
 Author: Joshua Ballanco<br />
Advisor: Marc Mansfield<br />
Date: May 9th, 2011<br />
Department: Chemistry, Chemical Biology, and Biomedical Engineering<br />
Degree: Doctor of Philosophy<br />
</div>
<!--l. 137--><p class="noindent" >


       <h2 class="likechapterHead"><a 
 id="x1-1000"></a>Table of Contents</h2> <div class="tableofcontents">
       <span class="chapterToc" >1 <a 
href="#x1-40001" id="QQ2-1-4">Introduction</a></span>
<br />       <span class="chapterToc" >2 <a 
href="#x1-70002" id="QQ2-1-8">Design of the Polymerase Evolution Model</a></span>
<br />       <span class="chapterToc" >3 <a 
href="#x1-130003" id="QQ2-1-15">Experimental Procedure and Results of Polymerase Modeling</a></span>
<br />       <span class="chapterToc" >4 <a 
href="#x1-160004" id="QQ2-1-33">Analysis and Discussion of Model Results</a></span>
<br />       <span class="appendixToc" >A <a 
href="#x1-21000A" id="QQ2-1-39">Source Code</a></span>
       </div>

       <h2 class="likechapterHead"><a 
 id="x1-2000"></a>List   of   Tables</h2><div class="tableofcontents"><span class="lotToc" >3.1&#x00A0;<a 
href="#x1-130011">Determining   genome   length   scale
effects.</a></span><br /><span class="lotToc" >3.2&#x00A0;<a 
href="#x1-130032">Determining size scale effects.</a></span><br /><span class="lotToc" >3.3&#x00A0;<a 
href="#x1-130053">Growth dynamics
at  various  temperatures.</a></span><br /><span class="lotToc" >3.4&#x00A0;<a 
href="#x1-140014">Competitive  growth  at  various
temperatures.</a></span><br /><span class="lotToc" >3.5&#x00A0;<a 
href="#x1-150015">Competitive growth in a full environment at
various temperatures.</a></span><br /><span class="lotToc" >3.6&#x00A0;<a 
href="#x1-150046">Competitive growth in a full environment
at many finely divided temperatures.</a></span><br />
       </div>

       <h2 class="likechapterHead"><a 
 id="x1-3000"></a>List of Figures</h2><div class="tableofcontents"><span class="lofToc" >1.1&#x00A0;<a 
href="#x1-50011"><span 
class="cmbx-12">Schematic of the Known </span>5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span><span 
class="cmbx-12">and the</span>
<span 
class="cmbx-12">Proposed </span>3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span><span 
class="cmbx-12">Polymerization Reactions. </span>In each panel, the incoming
nucleotide is colored blue and the pyrophosphate leaving-group is colored
red. The 3&#8217; and 5&#8217; carbon atoms on the sugars of the incoming nucleotide
and the terminal nucleotide of the growing chain are labeled (R represents
the remainder of the growing polymer). Note that the pyrophosphate
leaving-group is found on the incoming nucleotide in the 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>reaction,
but the leaving-group in the 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>reaction is on the growing chain
itself.</a></span><br /><span class="lofToc" >2.1&#x00A0;<a 
href="#x1-120011"><span 
class="cmbx-12">A schematic diagram of geometry based polymerase</span>
<span 
class="cmbx-12">discrimination. </span>The simplistic model of geometry based discrimination,
and its relationship to polymerase rate, assumes that a tighter binding
polymerase will be better able to exclude incorrect nucleotides based on
shape. However, being tighter binding will restrict the rate at which the
polymerase can translocate along the nucleic acid.</a></span><br /><span class="lofToc" >3.1&#x00A0;<a 
href="#x1-130021"><span 
class="cmbx-12">Effects of Genome</span>
<span 
class="cmbx-12">Length on Polymerase Rate Evolution. </span>The average polymerase rate
for the entire simulation population is plotted against simulation time steps
(each unit on the abscissa is equivalent to 100 simulation time steps). The
organisms each had genomes of length 10, 100, 1000, or 10000.</a></span><br /><span class="lofToc" >3.2&#x00A0;<a 
href="#x1-130042"><span 
class="cmbx-12">Effects of</span>
<span 
class="cmbx-12">Environmental Carrying Capacity on Polymerase Rate Evolution.</span>
The average polymerase rate for the entire simulation population is plotted
against simulation time steps (each unit on the abscissa is equivalent to 100

simulation time steps). The organisms each had a genome length of 1000, and
the environments had a maximum capacity (<span 
class="cmmi-12">N</span>) of 100, 1000, 10000, or 100000
organisms.</a></span><br /><span class="lofToc" >3.3&#x00A0;<a 
href="#x1-130063"><span 
class="cmbx-12">Exponential growth at various temperatures in the</span>
<span 
class="cmbx-12">absence of competition, with mutations. </span>The model system was seeded
with environments, at simulation temperatures of 0.10, 0.30, 0.40, or 0.60,
containing 10 organisms with a 5.5 average polymerase rate. <span 
class="cmbx-12">A</span>. Population
size of model organisms as a function of simulation time. <span 
class="cmbx-12">B</span>. Evolution of
the average polymerase rate for the organisms in each environment as a
function of simulation time. In each case, solid lines are used to indicate
environments with forward polymerizing organisms and dashed lines are for
reverse polymerizing organisms. Different temperatures are indicated with
different data markers as indicated in the figure legend, and are expressed
in units of <span 
class="cmr-8">&#x0394;</span><span 
class="cmmi-8">E</span> 
 <span 
class="cmmi-8">R</span> . </a></span><br /><span class="lofToc" >3.4&#x00A0;<a 
href="#x1-130074"><span 
class="cmbx-12">Exponential growth at various temperatures in</span>
<span 
class="cmbx-12">the absence of competition, with no mutations. </span>The model system
was seeded with environments, at simulation temperatures of 0.10, 0.30,
0.40, or 0.60, containing 10 organisms with a 5.5 average polymerase rate.
<span 
class="cmbx-12">A</span>. Population size of model organisms as a function of simulation time.
<span 
class="cmbx-12">B</span>. Evolution of the average polymerase rate for the organisms in each
environment as a function of simulation time. In each case, solid lines are
used to indicate environments with forward polymerizing organisms and
dashed lines are for reverse polymerizing organisms. For every simulation
mutations were disallowed. Different temperatures are indicated with
different data markers as indicated in the figure legend, and are expressed
in units of <span 
class="cmr-8">&#x0394;</span><span 
class="cmmi-8">E</span> 
 <span 
class="cmmi-8">R</span> . </a></span><br /><span class="lofToc" >3.5&#x00A0;<a 
href="#x1-140025"><span 
class="cmbx-12">Competition during exponential growth at</span>
<span 
class="cmbx-12">various temperatures, with mutations. </span>The model system was seeded
with environments, at simulation temperatures of 0.10, 0.30, 0.40, or 0.60,

containing 100 organisms, 50 each with forward and reverse polymerases,
with a 5.5 average polymerase rate. <span 
class="cmbx-12">A</span>. Population size of model organisms as
a function of simulation time. <span 
class="cmbx-12">B</span>. Evolution of the average polymerase rate for
the organisms in each environment as a function of simulation time. In each
case, solid lines are used to indicate environments with forward polymerizing
organisms and dashed lines are for reverse polymerizing organisms. Different
temperatures are indicated with different data markers as indicated in the
figure legend, and are expressed in units of <span 
class="cmr-8">&#x0394;</span><span 
class="cmmi-8">E</span> 
 <span 
class="cmmi-8">R</span> . </a></span><br /><span class="lofToc" >3.6&#x00A0;<a 
href="#x1-140036"><span 
class="cmbx-12">Competition during</span>
<span 
class="cmbx-12">exponential growth at various temperatures, with no mutations.</span>
The model system was seeded with environments, at simulation temperatures
of 0.10, 0.30, 0.40, or 0.60, containing 100 organisms, 50 each with forward
and reverse polymerases, with a 5.5 average polymerase rate. <span 
class="cmbx-12">A</span>. Population
size of model organisms as a function of simulation time. <span 
class="cmbx-12">B</span>. Evolution of
the average polymerase rate for the organisms in each environment as a
function of simulation time. In each case, solid lines are used to indicate
environments with forward polymerizing organisms and dashed lines
are for reverse polymerizing organisms. For every simulation mutations
were disallowed. Different temperatures are indicated with different data
markers as indicated in the figure legend, and are expressed in units of <span 
class="cmr-8">&#x0394;</span><span 
class="cmmi-8">E</span> 
 <span 
class="cmmi-8">R</span> .
</a></span><br /><span class="lofToc" >3.7&#x00A0;<a 
href="#x1-150027"><span 
class="cmbx-12">Competition in an environment at maximum capacity, with</span>
<span 
class="cmbx-12">mutations. </span>Environments, at simulation temperatures of 0.10, 0.30, 0.40,
or 0.60, were seeded with 500 organisms containing forward polymerases
and 500 containing reverse, both with a 5.5 average polymerase rate.
<span 
class="cmbx-12">A</span>. Population size of model organisms as a function of simulation time.
<span 
class="cmbx-12">B</span>. Evolution of the average polymerase rate for the organisms in each
environment as a function of simulation time. In each case, solid lines are

used to indicate environments with forward polymerizing organisms and
dashed lines are for reverse polymerizing organisms. Different temperatures
are indicated with different data markers as indicated in the figure legend,
and are expressed in units of <span 
class="cmr-8">&#x0394;</span><span 
class="cmmi-8">E</span> 
 <span 
class="cmmi-8">R</span> . </a></span><br /><span class="lofToc" >3.8&#x00A0;<a 
href="#x1-150038"><span 
class="cmbx-12">Competition in an environment at</span>
<span 
class="cmbx-12">maximum capacity, with no mutations. </span>Environments, at simulation
temperatures of 0.10, 0.30, 0.40, or 0.60, were seeded with 500 organisms
containing forward polymerases and 500 containing reverse, both with a
5.5 average polymerase rate. <span 
class="cmbx-12">A</span>. Population size of model organisms as a
function of simulation time. <span 
class="cmbx-12">B</span>. Evolution of the average polymerase rate for
the organisms in each environment as a function of simulation time. In each
case, solid lines are used to indicate environments with forward polymerizing
organisms and dashed lines are for reverse polymerizing organisms. For
every simulation mutations were disallowed. Different temperatures are
indicated with different data markers as indicated in the figure legend,
and are expressed in units of <span 
class="cmr-8">&#x0394;</span><span 
class="cmmi-8">E</span> 
 <span 
class="cmmi-8">R</span> . </a></span><br /><span class="lofToc" >3.9&#x00A0;<a 
href="#x1-150059"><span 
class="cmbx-12">Competition in an environment</span>
<span 
class="cmbx-12">at maximum capacity at various temperatures. </span>Simulations were
carried out with environments at temperatures ranging from 0.10 to 0.60
in 0.05 increments. In each simulation, the environment was seeded at
full capacity with 500 organisms containing forward polymerases and 500
containing reverse. For each variety there were 50 organisms with each of
the possible polymerase rates, giving an average rate of 5.5 In <span 
class="cmbx-12">A </span>and <span 
class="cmbx-12">B </span>the
natural log of the population of organisms containing reverse polymerases
is plotted as a function of simulation time. <span 
class="cmbx-12">A </span>is the data from simulations
where mutation was allowed, and <span 
class="cmbx-12">B </span>is from simulations where mutations
were not permitted. A plot of the slopes of a least squares regression line
for the data in each simulations is plotted as a function of simulation

temperature. Data from simulations with mutation is plotted as the solid
line with data from the no mutation simulations plotted as a dashed
line. </a></span><br /><span class="lofToc" >4.1&#x00A0;<a 
href="#x1-170011"><span 
class="cmbx-12">Generational change in polymerase rate as a function</span>
<span 
class="cmbx-12">of temperature. </span>The organisms from the systems plotted in 3.3 were
analyzed for the difference between the rate of their polymerase and the
rate of their parent&#8217;s polymerase. This difference is plotted for the various
different temperatures simulated. The maximal difference we would expect
to see (i.e. in the case that inheritance was purely random) would be 5. </a></span><br />
       </div>
<!--l. 168--><p class="indent" >



       <h2 class="chapterHead"><span class="titlemark">Chapter&#x00A0;1</span><br /><a 
 id="x1-40001"></a>Introduction</h2>
<!--l. 3--><p class="noindent" >Evolution has classically been a science of the past. That is, the primary sources of
evidence used in the formulation and validation of evolutionary theory have been the fossil
record, geology, and observations of the current form and distribution of species
as a result of events that took place in the past. Where evolution does exhibit
predictive power, it is primarily in the ability to predict what new evidence from
the past should eventually be revealed, such as the many fossils of transitional
forms discovered over the past centuries. In this regard, evolution has been wildly
successful.
<!--l. 5--><p class="indent" >       Where the study of Evolution has been thus far lacking is in its ability to
predict the future. That is, looking at a collection of fossils and historical records of
environmental conditions, evolutionary theory gives us the ability to identify which
selective pressures acted on past populations. What evolutionary theory cannot
do, at present, is predict which environmental conditions or physical forms will
act as selective pressures in the future. At best, we can make educated guesses
based on past evidence, but we lack the ability to assign a concrete measure of
confidence in such predictions. In the end, the only sure way to determine which
individuals are best fit is to wait and see which individuals ultimately survive and
reproduce.
<!--l. 7--><p class="indent" >       This is not an inherent failing of evolution, but rather a sign of a young and
developing field. The inability to predict the future course of evolution is tied directly to
the fact that selective pressures can come from practically any feature of an organism.
Everything from diet to behavior to physical form can lead to a selection for or against the
eventual reproductive success of an organism. The situation in which evolution
finds itself is akin to what physics would be like if every projectile had its own
unique laws of motion. What evolution desperately needs is a fundamental set

of principles from which the success of any organism can be derived, much as
the path of every projectile can ultimately be derived from Newton&#8217;s laws of
motion.
<!--l. 9--><p class="indent" >       The implications of a coherent method for deriving the evolutionary success of an
individual undergoing natural selection would be far reaching. Perhaps the most obvious
implication would be an improved ability to predict patterns of disease emergence and
transmission. Cancer treatment would be another potential area of benefit, as cancer is an
evolutionary process. With each round of chemotherapy or radiation treatment, those
cancer cells which are better adapted to resist treatment will survive. These cells
will then seed a reemergent, more difficult to treat, tumor. Understanding the
dynamics of evolution, and how to predict the future course of evolution, would
improve our ability to design effective treatments for both microbial diseases and
cancer.
<!--l. 11--><p class="indent" >       Even non-biological processes are driven by evolution and obey many of the same
laws that Darwin first laid out 150 years ago. Both languages and economies undergo
evolutionary processes, driven by the same math as cancer or the origin of species<span class="cite">[<a 
href="#XANowak:2006p857">10</a>]</span>.
That said, we must make a distinction between biological and non-biological
evolution. The reason for doing so lies in the approaches that can be taken to
investigate each type of evolutionary system. We know far more about biological
organisms than merely that they evolve. The past 50 years has seen an explosion
of knowledge about the chemistry of life and the operation of those molecular
systems which compose cells and influence the phenotypes of complex multicellular
organisms.
<!--l. 13--><p class="indent" >       For this reason, biological systems undergoing evolution present a unique
opportunity to attempt to derive the fundamental principles of evolution from the
individual mechanics of the evolving systems. Specifically, with biology we can explore the

feedback mechanism which drives evolution: the &#8220;Central Dogma&#8221; of biology. This is the
pathway by which information flows from an organism&#8217;s nucleic acids, where it is stored, to
the organism&#8217;s proteins which then drive fundamental biochemical processes<span class="cite">[<a 
href="#XThieffry:1998p913">14</a>]</span>. These
biochemical processes determine the phenotypes eventually selected for or against in the
process of natural selection. From this flow of information, it is understood that those
organisms which contain the information for generating the better fit traits will become
enriched in the population, and therefore will pass on this information in greater numbers
than their less fit peers<span class="cite">[<a 
href="#XDarwin:1883p935">4</a>]</span>.
<!--l. 15--><p class="indent" >       The biochemical processes that ultimately result in phenotypic traits can be very
complex, and their ultimate effect may be indirect. This makes it difficult to study the
biochemical evolution of all but the simplest of traits in the simplest of organisms. One
biochemical process that is relatively simple, and whose impact has a very direct effect on
the eventual transmission of genetic information, is the process of nucleotide synthesis.
Since a prime requirement for reproduction, especially in an era when all life was
unicellular, is the ability to replicate genetic information, one would expect this to be a
process under heavy selective pressure. Also, since the phenotype of reproduction rate is
very closely tied to and dependent upon the specific molecular mechanisms, this process
provides a prime opportunity to study how organism-scale selection manifests itself at the
molecular scale.
       <h3 class="likesectionHead"><a 
 id="x1-50001"></a>Nucleic Acid Polymerization</h3>
<!--l. 19--><p class="noindent" >The two classes of biologically important nucleic acids are Ribonucleic Acid (RNA) and
Deoxyribonucleic Acid (DNA). These two classes are very similar, differing only by the
presence or absence, respectively, of a 2&#8217; hydroxyl group on the ribose sugar of their
individual monomers and the specific nitrogenous bases attached at the 1&#8217; position. Both
types of polymer are formed by a process of dehydration synthesis catalyzed by a class of

enzymes known as nucleotide polymerases. The dehydration reaction takes place
between one of the hydroxyl groups on the <span 
class="cmmi-12">&#x03B1;</span>-phosphate of a nucleotide triphosphate
and the 3&#8217; hydroxyl group of the terminal monomer on the growing nucleic acid
chain<span class="cite">[<a 
href="#XZannisHadjopoulos:1999p1929">15</a>]</span>.
<!--l. 21--><p class="indent" >       What is phenomenal about this process is that, first, it occurs in all biological
organisms and, second, that it always occurs in the same manner. To understand why the
consistency of directionality is notable, it helps to understand the catalytic process
that occurs at the active center of nucleotide polymerases. All known nucleotide
polymerases share a common chemical mechanism. In this mechanism two divalent metal
cations, coordinated by a number of acidic amino acids, facilitate the transfer
of an electron pair from the free 3&#8217; hydroxyl group of one nucleotide to the <span 
class="cmmi-12">&#x03B1;</span>
phosphate, which is attached to the 5&#8217; hydroxyl of the other nucleotide<span class="cite">[<a 
href="#XCastro:2009p953">3</a>]</span> (see
Figure&#x00A0;<a 
href="#x1-50011">1.1<!--tex4ht:ref: fig:reactions --></a>).
<!--l. 23--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-50011"></a>


<!--l. 25--><p class="noindent" ><img 
src="reactions.png" alt="PIC"  
width="433.62pt" height="207.07144pt" >
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;1.1: </span><span  
class="content"><span 
class="cmbx-12">Schematic of the Known </span>5<span 
class="cmsy-10x-x-120">&#x2032; &#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span><span 
class="cmbx-12">and the Proposed </span>3<span 
class="cmsy-10x-x-120">&#x2032; &#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032;</span>
<span 
class="cmbx-12">Polymerization Reactions. </span>In each panel, the incoming nucleotide is colored blue
and the pyrophosphate leaving-group is colored red. The 3&#8217; and 5&#8217; carbon atoms on
the sugars of the incoming nucleotide and the terminal nucleotide of the growing
chain are labeled (R represents the remainder of the growing polymer). Note that
the pyrophosphate leaving-group is found on the incoming nucleotide in the 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>
reaction, but the leaving-group in the 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>reaction is on the growing chain itself.</span></div><!--tex4ht:label?: x1-50011 -->

<!--l. 28--><p class="indent" >       </div><hr class="endfigure">
<!--l. 30--><p class="indent" >       Importantly, the catalytically active cations of the polymerase active site have no
knowledge of whether the 5&#8217;-phosphate is attached to the incoming nucleotide or the
growing chain. If one were to imagine a nucleotide polymerase that proceeded in the reverse
direction of all currently known nucleotide polymerases, the only aspect of the
naturally occurring enzymes that would need to be modified are those portions
which attach to the growing nucleotide polymer or the incoming nucleotide, all of
which are distinct from the catalytic center. That is, there should be nothing
preventing the chemical reaction at a hypothetical 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase active site from
occurring.
<!--l. 32--><p class="indent" >       So the question remains: why do we not currently observe such 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerases
in nature? There are three explanations one might imagine for why no alternative
polymerases are found in nature: chemical impossibility, founder effect, or evolutionary
selection. Taking the first explanation, it may be that the chemistry involved in synthesis in
the reverse direction, 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032;</span>, is impossible, and only the known 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerization
reaction can be performed by biological enzymes. Yet, as noted above, catalysis does not
involve segments of the growing polymer or the nucleotide monomer other than the 3&#8217;
hydroxyl and <span 
class="cmmi-12">&#x03B1;</span>-phosphate. If the triphosphate group were attached to the growing chain
instead, and the 3&#8217; hydroxyl group were positioned on the incoming nucleotide
monomer, the active site configuration would not look any different. It is also
notable that the change in entropy of the reaction would be the same regardless
of polymerization reaction, as the leaving group is a pyrophosphate in either
case.
<!--l. 34--><p class="indent" >       The second possibility is that the unique directionality is the consequence of a
founder effect. A founder effect is when a small subset of a larger population is
evolutionarily isolated and goes on to give rise to a new population. This new population

would be expected to contain an oversampling of the traits of the small founding
population as compared to the gene frequencies of the original population<span class="cite">[<a 
href="#XTempleton:1980p954">13</a>]</span>. In the case
of nucleotide polymerization, this would imply that at some point early on in the
development of life on earth the population contained both forms of nucleotide
polymerase, those that polymerize by extension 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>and 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>. Then, at some
later point a subgroup of this population, consisting entirely of organisms with a
5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase, was isolated and subsequently gave rise to all life on earth
today.
<!--l. 36--><p class="indent" >       The other possibility is that there is an evolutionary advantage to the 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>
directionality, and that all life polymerizes nucleotides in this direction as the result of a
selection event. Unfortunately, the chances of finding any evidence directly supporting the
hypothesis of a founder effect determining polymerase directionality are essentially nil.
Short of finding some remnant modern population with reversed polymerases, the only
evidence of an ancient 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase would be the enzymes or other early replicator
molecules themselves, and individual molecules do not fossilize. The only way to shed light
on which of these two competing hypotheses explains the current state of nucleotide
polymerases would be to identify an evolutionary advantage imparted by a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>
polymerase over the reverse.
<!--l. 38--><p class="indent" >       What might be the source of such an advantage? Looking again at the chemical
reactions that would be carried out in each case (fig.&#x00A0;<a 
href="#x1-50011">1.1<!--tex4ht:ref: fig:reactions --></a>), a major difference between the
5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>case and the 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>case is the location of the pyrophosphate leaving-group
relative to the other components of the polymerization reaction. This is important because
it is known that triphosphate groups attached to nucleotides can spontaneously
hydrolyze<span class="cite">[<a 
href="#XSigel:1983p720">12</a>]</span>. In the event that the triphosphate leaves prematurely with the 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>
polymerase, it would be possible for the enzymatic machinery to discard the (now useless)
incoming nucleotide and replace it with a new nucleotide containing an active triphosphate.

In the event of a spontaneous hydrolysis with the 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase, because the active
triphosphate in the reaction is on the growing chain the only choice the enzymatic
machinery has is to discard all of its progress thus far or wait for some secondary
machinery to replace the triphosphate group. Given that the former of these two choices
represents significant waste, we can assume that the later choice is the only one that
would be even remotely sustainable, but this choice involves introducing a delay in
reproduction.
<!--l. 40--><p class="indent" >       A delay, on its own, does not necessarily mean that the 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>case is at enough of
a disadvantage to be completely eliminated from a population through natural selection.
After all, polymerase rates are variable and a hypothetical 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase could simply
evolve to polymerize faster. Unfortunately, such modifications come with a cost. There is a
relationship between speed of nucleotide polymerization and error rate<span class="cite">[<a 
href="#XGriep:2006p719">6</a>]</span>, and polymerizing
faster would cause the polymerase to make more mistakes. A higher frequency of errors
during reproduction would introduce a higher rate of mutation, making it more difficult for
the 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerizing organisms to maintain their speed advantage. Again, on its
own this is not reason enough to dismiss the possibility of a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase.
There are many factors at play, and to understand how all of these combined
contribute to the fitness of an organism that synthesizes its nucleic acids in a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032;</span>
direction, we will need to model the competitive evolution of the two polymerase
types.
<!--l. 42--><p class="indent" >       Finally, it is easy enough to imagine a polymerase which operates in much the same
fashion as modern nucleotide polymerases, but uses as its substrate nucleotides with
3<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate moieties in place of nucleotide-5<span 
class="cmsy-10x-x-120">&#x2032;</span>-phosphates and does not incur an excess
penalty due to spontaneous hydrolysis of the triphosphate group. This hypothetical
situation can be discounted, however, by considering that ribose-3-phosphate is known to
decompose more readily under mildly acidic conditions than ribose-5-phosphate<span class="cite">[<a 
href="#XLevene:1934p869">9</a>]</span>,

consistent with the availability of a nucleophilic oxygen on the adjacent carbon at position
2 in ribose-3-phosphate. The implication is that the primordial pool of nucleotides would
lack sufficient quantities of nucleotide-3<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphates with which to do synthesis. So,
we can assume that any reverse polymerase would have to work with the same
nucleotide-5<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphates as a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase, and therefore must face the potential
penalty of doing so.
       <h3 class="likesectionHead"><a 
 id="x1-60001"></a>Modeling Evolution</h3>
<!--l. 47--><p class="noindent" >Because nucleotide polymerases are responsible for replicating the genetic information of an
organism and have a direct influence on both the rate at which an organism can reproduce
as well as the fidelity with which genetic information is conveyed from one generation to
the next, the approach we take in simulating this process is relatively straightforward. The
model presented here involves simplified model organisms designed to represent the earliest
forms of self-replicating life. It is assumed these organisms exist in an environment rich
with nutrients and an excess of available energy. The result of this being that the
rate of genome duplication serves as the lone limiting factor on reproductive
rate.
<!--l. 49--><p class="indent" >       Since polymerase directionality would have been fixed extremely early on in the
origin of life, it is reasonable to assume that the recombination of genetic elements would
contribute only a negligible amount of variation to the model organisms. Additionally, the
simplest of nucleotide polymerases observed in nature are the products of single genes.
Even if there was recombination of individual genetic elements, the exclusive focus of this
model is the polymerase. Whether a product polymerase gene is transferred intact
to the offspring of one organism or transferred horizontally will not affect the
conclusions that can be made. This is a result of the first assumption made that

the model organisms have an excess of all resources aside from the polymerase.
That is, whatever organism a polymerase may end up in, the polymerase will
always be the single determining factor for both reproductive rate and genetic
fidelity.
<!--l. 51--><p class="indent" >       Ultimately, the goal of this model is to investigate the link between a well
understood molecular process, nucleotide polymerization, and an evolutionary outcome, the
selection for polymerase enzymes which proceed 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>. If there are general rules to be
found that govern the process of evolution all the way from the molecular level to the scale
of entire organisms and even populations, hopefully starting with a simple model
like the one presented here will serve as an important first step in their eventual
discovery.



       <h2 class="chapterHead"><span class="titlemark">Chapter&#x00A0;2</span><br /><a 
 id="x1-70002"></a>Design of the Polymerase Evolution Model</h2>
<!--l. 3--><p class="noindent" >This chapter provides a detailed description of the design of the model system used to
investigate the question of the evolution of nucleotide polymerase directionality. The goal of
this model is to have organisms containing either a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>or a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>nucleic acid
polymerase compete with each other in order to see which of the strategies, if either, are
evolutionarily dominant. In the model, evolutionary pressure is applied by the rate at
which organisms can reproduce and the fidelity with which they duplicate their genomes.
Evolutionary momentum is introduced through simple genetic mutation, which could be
enabled or disabled on a per experiment basis. The model is also designed such
that the influence of temperature on the outcome of this competition can be
investigated.
       <h3 class="likesectionHead"><a 
 id="x1-80002"></a>Overview</h3>
<!--l. 7--><p class="noindent" >In simplest terms, this model consists of a number of individual organisms in competition
with each other. The model can be largely divided up into, and thought about most clearly
as, four interacting components. These are the environment, the organisms, the genomes,
and the polymerases. The relationship between these components begins with the
environment. Each run of the simulation involves exactly one environment. An environment
may contain many organisms, up to a predefined carrying capacity, and at the beginning of
a simulation run the environment is pre-populated organisms according to a set
rules.
<!--l. 9--><p class="indent" >       Each organism contains one genome. At the start of an organisms &#8220;life&#8221;, it uses its
genome to create a polymerase. The polymerase replicates either 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>or 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032;</span>
and may not change direction. The polymerase then replicates the genome at a
predetermined rate and with a random chance of introducing errors (mutations). When
the polymerase is finished duplicating the genome, the organism will attempt

to divide by binary fission using the newly created genome. If it is successful
in doing so, then the new organism is added to the environment and begins a
fresh life cycle while the original organism resets itself and also begins a fresh
cycle.
<!--l. 11--><p class="indent" >       A number of generalizations and assumptions have been made in order to render the
problem being investigated tractable, but a consistent effort has been made to remain as
true to life as possible. The motivation behind this is that, while the exact values generated
by this model may not be precisely those that would be observed in the laboratory, the
trends observed should hold true when transferred to the bench. For each piece decisions
have been made regarding which aspects of the component are explored in depth, and
which aspects are neglected, with an eye to the larger goal of investigating the role
of polymerase directionality. Following is a detailed look at each component in
turn.
<!--l. 14--><p class="noindent" >
       <h3 class="likesectionHead"><a 
 id="x1-90002"></a>Environment</h3>
<!--l. 16--><p class="noindent" >At the start of a simulation run, the environment for that run is initialized with a starting
population of organisms. To create this starting population, a set of organisms with their
genome length, polymerase rate, and polymerase directionality specified is divided
according each organism&#8217;s designated frequency in the starting population. The starting
population size can be any number up to the specified maximum population of the
environment.
<!--l. 18--><p class="indent" >       A key defining feature of this model is that the environment is constrained in some
ways, but not in others. Specifically, the number of organisms that can simultaneously
co-exist has a hard numerical limit; a carrying capacity. On the other hand, the amount of

available energy, the quantity of activated nucleotide triphosphates, and the other raw
materials required for forming a cell are all considered unlimited. In reality, the carrying
capacity is a generalization that could correspond to physical space limits, but it could just
as easily correspond to the aggregate limitations on the other resources not explicitly
modeled.
<!--l. 20--><p class="indent" >       The choice of limitation based on carrying capacity was driven by two factors. The
first relates to the fact that selection during an exponential growth phase could potentially
operate differently than during stationary phase growth. By placing an upper limit on the
size of a population, it is possible to investigate selection during both growth phases.
Second, while attempting to explicitly model all of the various resource and space
constraints that might ultimately limit growth might yield a more complete model, it
would also significantly increase the complexity of the model without adding much insight
into the question at hand. Finally, numerically or density limited growth is a common
observation across a wide range of living creatures from single cells to large populations of
complex animals<span class="cite">[<a 
href="#XFerenci:1999p703">5</a>]</span>.
<!--l. 22--><p class="indent" >       In order to model density dependent growth inhibition as the number of organisms
approaches the carrying capacity, a random death probability is introduced to the
environment. In keeping with observations that the pressure of density dependent
inhibition is greater as the population of an environment approaches the carrying capacity,
the death probability is calculated with an inverse function of the remaining
capacity
       <center class="math-display" >
<img 
src="steventhesis_num0x.png" alt="              1
Pdeath =  ------------
         (N  - n ) + 1
" class="math-display" ></center>

<!--l. 24--><p class="nopar" >where <span 
class="cmmi-12">N </span>is the carrying capacity, <span 
class="cmmi-12">n </span>is the number of organisms currently in the
environment, and 1 is added so that the probability of at least one organism being culled
from the environment when the carrying capacity is reached is <span 
class="cmmi-12">P</span><sub><span 
class="cmmi-8">death</span></sub> = 1.
<!--l. 26--><p class="indent" >       The model iterates its environment in a stepwise fashion. During each step, each of
the organisms contained within the environment is allowed to carry out one time-step of its
life cycle, followed by a population culling. Culling is carried out by calculating the death
probability as above, and then randomly applying that probability to the environment to
decide if an organism should be removed. If the decision is made to remove an
organism, one is chosen from the environment randomly, removed, and the death
probability is recalculated and reapplied. This process will repeat until the decision is
made not to remove an organism. In this way, the number of organisms removed
at each time step follows a Poisson distribution with an expectation value of
<span 
class="cmmi-12">P</span><sub><span 
class="cmmi-8">death</span></sub>.
<!--l. 29--><p class="noindent" >
       <h3 class="likesectionHead"><a 
 id="x1-100002"></a>Organism</h3>
<!--l. 31--><p class="noindent" >Organisms are created and added to the environment either at the start of a simulation
run, as part of the starting population, or during the simulation run as the result of
binary fission of an existing organism. If the organism is created at the start of the
simulation then its genome and polymerase properties are determined by the
values used to seed the population. Each of these seed organisms is generated with
a random amount of its genome already synthesized, to avoid synchronization
artifacts that arise if all organisms start polymerizing from zero at the start of the
simulation.
<!--l. 33--><p class="indent" >       If the organism is created during a simulation run, then its properties are derived

from those of its parent, with the possibility of introduced variation. This variation is
embodied by, and in the model determined by, the genome contained within the
organism. Each organism contains exactly one genome and one polymerase. Having
only one polymerase is a simplification, but it is justified by considering this one
polymerase as an exemplar of all the polymerase enzymes that would be found in a real
organism.
<!--l. 35--><p class="indent" >       Each organism is modeled as a state machine. The two states in which an organism
can exist are: <span 
class="cmti-12">Polymerizing </span>or <span 
class="cmti-12">Duplicating</span>. At creation, each organism starts in the
<span 
class="cmti-12">Polymerizing </span>state. When an organism is in the <span 
class="cmti-12">Polymerizing </span>state, each simulation
time-step is used to allow the organism&#8217;s polymerase to add nucleotides to the genome copy
being constructed. At the end of each time-step, the polymerase is queried as to
whether or not it is finished with constructing the nascent genome. If it is, then the
organism shifts to the <span 
class="cmti-12">Duplicating </span>state. Otherwise, it remains in the <span 
class="cmti-12">Polymerizing</span>
state.
<!--l. 37--><p class="indent" >       When an organism reaches the <span 
class="cmti-12">Duplicating </span>state, a check is performed to determine
if the genome copy constructed by the polymerase is viable. This is a simple sanity check to
ensure that the synthesized genome is of the correct length. If the genome copy is not
viable, then it is discarded and the organism returns to the <span 
class="cmti-12">Polymerizing </span>state to create a
new genome copy. If the genome is viable, the next determination that must be
made is whether or not there is available capacity in the environment for a new
organism.
<!--l. 39--><p class="indent" >       If the number of organisms in the environment is not at the environment&#8217;s carrying
capacity at the start of a simulation time step, then an organism in the <span 
class="cmti-12">Duplicating </span>state
can proceed to create a new organism. The thus created organism containing the newly
synthesized genome is inserted into the environment. In this case, the parent organism will
return to the <span 
class="cmti-12">Polymerizing </span>state to begin construction of yet another new genome. If,

instead, the environment is at capacity at the start of a simulation time step, then any
organism in the <span 
class="cmti-12">Duplicating </span>state will remain unchanged, in the <span 
class="cmti-12">Duplicating </span>state. It will
next attempt to add a new organism with the synthesized genome during the next
environment step.
<!--l. 42--><p class="noindent" >
       <h3 class="likesectionHead"><a 
 id="x1-110002"></a>Genome</h3>
<!--l. 44--><p class="noindent" >Genomes are created before the organism in which they are contained. In the
case of a running simulation, the new genome is created by the parent organism
before binary fission. For the starting population of organisms used to seed a
simulation, the genomes are created ahead of time with predefined properties and a
random amount of replication already completed. The model genomes perform three
important functions for the simulation: they generate polymerases with properties
defined by the genome, they track the progress of the organism&#8217;s polymerase during
polymerization, and they return a copy of themselves once polymerization has
completed.
<!--l. 46--><p class="indent" >       Each genome codes for either a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase or a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase, and this
will not change through subsequent generations. There is the possibility that
this is not perfectly reflective of how early organisms may have functioned, as it
is conceivable that, through acquired mutations, a genome which coded for a
3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase may evolve in subsequent generations into a genome that codes
for a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>. Were this to occur, it would either have to be through a gradual
mechanism or a small mutation affecting global properties of the polymerase. A gradual
mechanism would be one where the forward or reverse speed of the polymerase is
reduced to nearly zero, and then eventually the directionality flipped. It is safe to

eliminate this possibility in the model, as such a transition would require the
survival of an organism with an especially slow polymerase. As a consequence of the
starting assumption that polymerase speed is the primary determining factor in
the speed of replication, such organisms would be quickly out-competed in the
simulation.
<!--l. 48--><p class="indent" >       The alternative mechanism of a small mutation leading to a global change can also
be neglected for two reasons. First, if such a mutation were possible, the fidelity of
communicating polymerase directionality from one generation to the next would be
diminished. This would imply that selecting against a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase would
effectively be impossible, as such polymerases could appear at almost any time. This
would also narrow the window of possible founder populations were the alternate
founder effect hypothesis considered. In the case that polymerase directionality were
trivially reversible, the hypothetical founder population would not only have to be
homogenous in their polymerase directionality, but they would also have to have a
unique restriction on the directionality reversing randomly. This also highlights the
second reason that the possibility of such a mutation need not be considered: if
such mutations were possible, it would be expected that they could be observed
today. Certainly, it is possible that such a mutation might have existed in the past
but have subsequently been locked out of the population of current polymerase
sequences, but this seems unlikely. Nucleotide polymerases are wildly variable in their
sequence and structure throughout all forms of life, and yet none exhibit reversed
directionality.
<!--l. 50--><p class="indent" >       When a genome is asked by its organism to create a polymerase, it will imbue the
polymerase with a fixed polymerase rate in addition to a fixed directionality. The precise
value the polymerase speed will take is determined when the genome is created from its
parent (or from predefined values at the beginning of a simulation). Once the

polymerase has been generated and polymerization begun, the polymerase will
inform the genome during each simulation time-step how many nucleotides have
been added to the copy being created, and how many of those are erroneous
base pairs. The genome keeps track of this information for later reporting and
calculations.
<!--l. 52--><p class="indent" >       When a genome copy is constructed, the genome can then provide that copy to a
new organism, but in doing so it must determine the properties of this new genome. As
mentioned above, the directionality is inherited exactly. The polymerase speed coded for by
the copy genome is determined in a two step calculation based on the number of errors
(mutations) introduced. First, the variance from the parent genome&#8217;s encoded polymerase
rate is calculated by taking
       <center class="math-display" >
<img 
src="steventhesis_num1x.png" alt="     rmax - rmin    M
&#x0394; =  ------------* ------
          2        Mmax
" class="math-display" ></center>
<!--l. 54--><p class="nopar" > where <span 
class="cmmi-12">r</span><sub><span 
class="cmmi-8">max</span></sub> and <span 
class="cmmi-12">r</span><sub><span 
class="cmmi-8">min</span></sub> are the maximum and minimum possible rates, respectively, <span 
class="cmmi-12">M </span>is the
number of errors made during replication, taken as a fraction of total genome length, and
<span 
class="cmmi-12">M</span><sub><span 
class="cmmi-8">max</span></sub> is the maximum tolerated fraction of errors, which was set at <img 
src="steventhesis_num2x.png" alt="1
3"  class="frac" align="middle">. The maximum
tolerated fraction of errors is the threshold above which it can be assumed that the
inherited traits of the polymerase are, essentially, selected at random. The maximum
and minimum rates were chosen as 1 and 10 nucleotides added per simulation
time-step to reflect the roughly 10-fold range of empirically observed polymerase
rates.
<!--l. 57--><p class="indent" >       The logic behind this metric is that more mutations will, on average, lead to
greater alteration of enzyme function. Certainly there are individual mutations

that could, on their own, have a large effect on polymerase rate, but averaged
over many polymerases and many possible mutations, this general trend should
hold. Once the variance is determined, it must be applied to the parent genome&#8217;s
encoded polymerase rate in such a way that the child&#8217;s genome does not code for an
invalid polymerase. So, if the variance would generate a polymerase rate above
the maximum, then it is subtracted from the parent rate, and vice versa. If the
variance would not result in an invalid rate as a consequence of either addition or
subtraction, then it is applied to the parent rate with an equal probability of either
outcome.
<!--l. 60--><p class="noindent" >
       <h3 class="likesectionHead"><a 
 id="x1-120002"></a>Polymerase</h3>
<!--l. 62--><p class="noindent" >The single most significant component of the model is the model nucleotide polymerase
itself. The other three components described above serve primarily to ensure that
selection and descent with modification will occur, but it is the polymerase which
embodies the trait that is ultimately being selected for or against: the polymerase&#8217;s
directionality. As with the genomes, the directionality and rate of a polymerase does
not change subsequent to its creation. These properties are determined by the
genome from which the polymerase is generated. The polymerases only have one
function during a simulation: to add nucleotides to a growing genome copy. How
this occurs depends on the polymerase directionality and the temperature of the
simulation.
<!--l. 64--><p class="indent" >       When an organism in the <span 
class="cmti-12">Polymerizing </span>state carries out one step of simulation
time, the polymerase will add nucleotides to the genome being synthesized. The
polymerase does this by running through a loop with a maximum number of

iterations determined by its rate. So, for example, a polymerase with a rate of
7 can carry out its nucleotide addition loop a maximum of 7 times for each 1
simulation time step. During each loop, two determinations must be made. First,
the polymerase must determine whether the nucleotide about to be added is
activated, and second it must determine whether a properly base-paired nucleotide is
being added, or whether the nucleotide being added will generate a replication
error.
<!--l. 66--><p class="indent" >       The reason that the polymerase must determine whether or not the nucleotide being
added is activated is due to a consequence of chemistry. All known nucleotide polymerases
in nature use, as their substrate, nucleotide triphosphates. They use the energy released by
cleaving the bond between the <span 
class="cmmi-12">&#x03B1; </span>and <span 
class="cmmi-12">&#x03B2; </span>phosphates attached to their 5<span 
class="cmsy-10x-x-120">&#x2032; </span>carbon to drive the
polymerase reaction forward. However, this bond can be cleaved through spontaneous
hydrolysis by water. The equilibrium between active and inactive nucleotides can be
described by a Boltzmann distribution:
       <center class="math-display" >
<img 
src="steventhesis_num3x.png" alt="       &#x0394;G0-
K =  e- RT
" class="math-display" ></center>
<!--l. 68--><p class="nopar" > where &#x0394;<span 
class="cmmi-12">G</span><sup><span 
class="cmr-8">0</span></sup> is the standard free energy of the hydrolysis reaction. To simplify the
simulation calculations, the entire <img 
src="steventhesis_num4x.png" alt="&#x0394;G0-
 RT"  class="frac" align="middle"> term is expressed as a generic simulation
temperature <img 
src="steventhesis_num5x.png" alt="1
t"  class="frac" align="middle">.
<!--l. 71--><p class="indent" >       Where this process of inactivation becomes important is in how it will differentially
affect a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase vs a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase. In the 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>case, the one
that is observed in all known forms of life, the triphosphate is attached to the
incoming nucleotide about to be added to the chain. If this incoming nucleotide

becomes inactive before the polymerase has a chance to join it to the growing
nucleic acid polymer, then this represents a wasted addition step. That is, if a
polymerase with a rate of 7 has one of its nucleotides become inactive before
phosphodiester bond formation, then it will only add 6 nucleotides during this simulation
time-step.
<!--l. 73--><p class="indent" >       On the other hand, in the case of a hypothetical 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase, the
triphosphate group would be attached to the 5<span 
class="cmsy-10x-x-120">&#x2032; </span>carbon of the nascent nucleic acid polymer.
If this triphosphate is cleaved during an addition step, the polymerase will not be
able to compensate by drawing in a new nucleotide from the pool of available
nucleotides. Rather, the polymerization process will necessarily halt while the
terminal end of the growing chain is reactivated. In the simulation, that process is
modeled by terminating the polymerase&#8217;s addition of new nucleotides when an
inactivation event occurs. That is, if a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase with a rate of 7 experiences
inactivation during the addition of the 3rd nucleotide in this time step, then only
2 new nucleotides will be added to the growing genome during this particular
step.
<!--l. 75--><p class="indent" >       Furthermore, whereas a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase will be constantly drawing in new,
activated nucleotides from the pool of all available nucleotides, a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase
depends on the nascent chain remaining activated. The consequence of this is that the
probability of a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase experiencing a deactivation event depends only on the
free energy of the reaction and temperature, irrespective of polymerase rate. Alternatively,
a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase runs a greater risk of the triphosphate on the nascent chain becoming
deactivated the longer it waits before adding a new nucleotide. So, for the reverse
polymerases, the distribution was adjusted by a factor determined by the polymerase
rate:
       <center class="math-display" >

<img 
src="steventhesis_num6x.png" alt="       1
K  = e- t * (rmax - r + 1 )
" class="math-display" ></center>
<!--l. 77--><p class="nopar" > where <span 
class="cmmi-12">r</span><sub><span 
class="cmmi-8">max</span></sub> is the maximum polymerase rate and <span 
class="cmmi-12">r </span>is the rate of the polymerase doing the
addition. This extra factor accounts for the fact that a slower reverse polymerase will give
each terminal phosphate group on the growing chain a longer period of time before addition
of the next nucleotide monomer during which a spontaneous hydrolysis might
occur.
<!--l. 79--><p class="indent" >       Finally, each time a new nucleotide is added by a polymerase, there is
a risk that the nucleotide will be of the wrong sort. The primary driving force
discriminating properly base-paired nucleotides from improperly paired nucleotides is the
formation of hydrogen bonds between the incoming nucleotide and the corresponding
nitrogenous base of the template nucleic acid. The hydrogen bond interaction
that allows discrimination between Watson-Crick base-pairing and a variety of
mismatches was modeled using the same sort of Boltzmann distribution as used for the
spontaneous hydrolysis condition. However, since the &#x0394;&#x0394;<span 
class="cmmi-12">G</span><sub><span 
class="cmr-8">0</span></sub> of a correct versus
incorrect base pairing is twice that of the spontaneous hydrolysis reaction<span class="cite">[<a 
href="#XPetruska:1988p721">11</a>,&#x00A0;<a 
href="#XBagshaw:1973p728">1</a>]</span>, the
hydrogen bond portion of the erroneous inclusion calculation is made using the
formula:
       <center class="math-display" >
<img 
src="steventhesis_num7x.png" alt="            - 2
PH -bond = e t
" class="math-display" ></center>
<!--l. 81--><p class="nopar" > where <span 
class="cmmi-12">t </span>is the simulation temperature.
<!--l. 84--><p class="indent" >       Recently, research has indicated that hydrogen bond formation is not sufficient to

completely explain the fidelity of correct pair formation by nucleotide polymerases. It has
been suggested that polymerases also take advantage of geometric differences between
proper and improper pair formation to improve their ability to discriminate correct versus
incorrect base pairing. Prior studies have revealed a relatively complex link between
polymerase rate and error rate for nucleotide polymerases<span class="cite">[<a 
href="#XCady:2009p718">2</a>]</span>. The essence of this
link is that there is an additional geometric constraint on erroneous nucleotide
inclusions and that, in order to polymerize faster, polymerases must relax this
constraint.
<!--l. 86--><p class="indent" >       In the model, this additional factor is treated as a simple flow problem. If a
polymerase is roughly represented as a cylindrical tube, the narrower the tube is
the greater will be its ability to reject geometrically unfavorable nucleotide pair
configurations. However, the rate of the polymerase will also be affected by the
radius of the tube (see figure&#x00A0;<a 
href="#x1-120011">2.1<!--tex4ht:ref: fig:schematic --></a> for an illustration of this concept). If rate is
taken as the flux of nucleic acid through the tube, then we end up with a simple
square root relationship between the ability of the polymerase to reject based on
geometry and the number of nucleotides it can add in one simulation time-step.
Specifically
       <center class="math-display" >
<img 
src="steventhesis_num8x.png" alt="     &#x221A; --
&#x0393;  =   r
" class="math-display" ></center>
<!--l. 88--><p class="nopar" > describes the geometric discrimination and
       <center class="math-display" >
<img 
src="steventhesis_num9x.png" alt="Perror = K * &#x0393;
" class="math-display" ></center>
<!--l. 91--><p class="nopar" > is the probability of making a mistake during each nucleotide addition event, calculated as
the product of the temperature dependent Boltzmann distribution, <span 
class="cmmi-12">K</span>, and the
geometric discrimination function, &#x0393;. Alternatively, in order to investigate the effect of
mutation on evolution, <span 
class="cmmi-12">P</span><sub><span 
class="cmmi-8">error</span></sub> could be artificially set to 0 at the start of a simulation
run.
<!--l. 94--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-120011"></a>


<!--l. 96--><p class="noindent" ><img 
src="schematic.png" alt="PIC"  
width="433.62pt" height="269.33334pt" >
<br />  <div class="caption" 
><span class="id">Figure&#x00A0;2.1:   </span><span  
class="content"><span 
class="cmbx-12">A   schematic   diagram   of   geometry   based   polymerase</span>
<span 
class="cmbx-12">discrimination. </span>The simplistic model of geometry based discrimination, and its
relationship to polymerase rate, assumes that a tighter binding polymerase will be
better able to exclude incorrect nucleotides based on shape. However, being tighter
binding will restrict the rate at which the polymerase can translocate along the nucleic
acid.</span></div><!--tex4ht:label?: x1-120011 -->

<!--l. 99--><p class="indent" >       </div><hr class="endfigure">



       <h2 class="chapterHead"><span class="titlemark">Chapter&#x00A0;3</span><br /><a 
 id="x1-130003"></a>Experimental Procedure and Results of Polymerase Modeling</h2>
<!--l. 3--><p class="noindent" >The procedure for performing an experiment with the model for polymerase directionality
evolution consists of constructing an environment for the model to operate on, then setting
the model running for a set number of time steps. Various parameters of the simulation are
reported at set intervals during the run. Because a number of the dynamical systems in the
model depend on randomly generated numbers, each experiment was carried out
in 10 copies to smooth fluctuations, and all of the values reported represent a
numerical average of all 10 runs. All plots were generated using the R software
package<span class="cite">[<a 
href="#XIhaka:1996p1018">7</a>]</span>.
<!--l. 5--><p class="indent" >       The dynamics included in the model for polymerase directionality should generate
scale free results. That is, the evolution of traits should show the same dynamics regardless
of population size or genome length. In this model, the only trait capable of evolving is
polymerase rate, so in order to verify that the dynamics were indeed scale free, the
first experiments carried out were designed to investigate simple dynamics of
polymerase rate evolution over a range of scales. One experiment consisted of starting
populations of 10 organisms and a maximum population (environmental carrying
capacity) of 1000 organisms combined with 4 different genome lengths as described in
table&#x00A0;<a 
href="#x1-130011">3.1<!--tex4ht:ref: tab:scale_length --></a>.
       <div class="table">

<!--l. 8--><p class="indent" >       <a 
 id="x1-130011"></a><hr class="float"><div class="float" 
>

<div class="center" 
>
<!--l. 8--><p class="noindent" >
<div class="tabular"> <table id="TBL-1" class="tabular" 
cellspacing="0" cellpadding="0" rules="groups" 
><colgroup id="TBL-1-1g"><col 
id="TBL-1-1"></colgroup><colgroup id="TBL-1-2g"><col 
id="TBL-1-2"></colgroup><colgroup id="TBL-1-3g"><col 
id="TBL-1-3"></colgroup><colgroup id="TBL-1-4g"><col 
id="TBL-1-4"></colgroup><tr  
 style="vertical-align:baseline;" id="TBL-1-1-"><td  style="white-space:nowrap; text-align:left;" id="TBL-1-1-1"  
class="td11">Temp. (<span 
class="cmmi-12">t</span>)</td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-1-2"  
class="td11">Starting Pop.</td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-1-3"  
class="td11">Max Pop.</td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-1-4"  
class="td11">Genome Length</td>
</tr><tr 
class="hline"><td><hr></td><td><hr></td><td><hr></td><td><hr></td></tr><tr  
 style="vertical-align:baseline;" id="TBL-1-2-"><td  style="white-space:nowrap; text-align:left;" id="TBL-1-2-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-2-2"  
class="td11">10               </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-2-3"  
class="td11">1000       </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-2-4"  
class="td11">10                  </td></tr><tr  
 style="vertical-align:baseline;" id="TBL-1-3-"><td  style="white-space:nowrap; text-align:left;" id="TBL-1-3-1"  
class="td11">0.40 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-3-2"  
class="td11">10 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-3-3"  
class="td11">1000 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-3-4"  
class="td11">100</td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-1-4-"><td  style="white-space:nowrap; text-align:left;" id="TBL-1-4-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-4-2"  
class="td11">10               </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-4-3"  
class="td11">1000       </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-4-4"  
class="td11">1000               </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-1-5-"><td  style="white-space:nowrap; text-align:left;" id="TBL-1-5-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-5-2"  
class="td11">10               </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-5-3"  
class="td11">1000       </td><td  style="white-space:nowrap; text-align:left;" id="TBL-1-5-4"  
class="td11">10000             </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-1-6-"><td  style="white-space:nowrap; text-align:left;" id="TBL-1-6-1"  
class="td11">         </td></tr></table></div>
<br /> <div class="caption" 
><span class="id">Table&#x00A0;3.1: </span><span  
class="content">Determining genome length scale effects.</span></div><!--tex4ht:label?: x1-130011 -->
</div>

       </div><hr class="endfloat" />
       </div>
<!--l. 22--><p class="indent" >       In order to avoid a founder bias with regards to polymerase rate, in each experiment
the ten organisms in each seed population consisted of one organism of each of the ten
possible polymerase rates. The simulation temperature of 0.40 was chosen because it results
in a significant amount of error, and therefore variability due to mutations, during growth
but is not a high enough temperature that certain other temperature effects begin
to alter growth and evolution. The results from this experiment are plotted in
figure&#x00A0;<a 
href="#x1-130021">3.1<!--tex4ht:ref: fig:scale_length --></a>.
<!--l. 24--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-130021"></a>


<!--l. 26--><p class="noindent" ><img 
src="scale_length.png" alt="PIC"  
width="433.62pt" height="261.17322pt" >
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;3.1:  </span><span  
class="content"><span 
class="cmbx-12">Effects  of  Genome  Length  on  Polymerase  Rate  Evolution.</span>
The average polymerase rate for the entire simulation population is plotted against
simulation time steps (each unit on the abscissa is equivalent to 100 simulation time
steps). The organisms each had genomes of length 10, 100, 1000, or 10000.</span></div><!--tex4ht:label?: x1-130021 -->

<!--l. 29--><p class="indent" >       </div><hr class="endfigure">
<!--l. 31--><p class="indent" >       Second, a set of experiments were carried out to investigate the effect of population
size on the evolution of polymerase rate. In this experiment, the genome length of the
organisms in each environment was held constant at 1000, the starting population was set
to 10, and the maximum population was set to either 100, 1000, 10000, or 100000. As in
the first set of experiments, the starting population in each environment was
seeded with organisms with polymerase rates evenly distributed in the range 1-10
and the simulation temperature was set to 0.40. Table&#x00A0;<a 
href="#x1-130032">3.2<!--tex4ht:ref: tab:scale_num --></a> summarizes these
experiments.
       <div class="table">

<!--l. 34--><p class="indent" >       <a 
 id="x1-130032"></a><hr class="float"><div class="float" 
>

<div class="center" 
>
<!--l. 34--><p class="noindent" >
<div class="tabular"> <table id="TBL-2" class="tabular" 
cellspacing="0" cellpadding="0" rules="groups" 
><colgroup id="TBL-2-1g"><col 
id="TBL-2-1"></colgroup><colgroup id="TBL-2-2g"><col 
id="TBL-2-2"></colgroup><colgroup id="TBL-2-3g"><col 
id="TBL-2-3"></colgroup><colgroup id="TBL-2-4g"><col 
id="TBL-2-4"></colgroup><tr  
 style="vertical-align:baseline;" id="TBL-2-1-"><td  style="white-space:nowrap; text-align:left;" id="TBL-2-1-1"  
class="td11">Temp. (<span 
class="cmmi-12">t</span>)</td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-1-2"  
class="td11">Starting Pop.</td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-1-3"  
class="td11">Max Pop.</td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-1-4"  
class="td11">Genome Length</td>
</tr><tr 
class="hline"><td><hr></td><td><hr></td><td><hr></td><td><hr></td></tr><tr  
 style="vertical-align:baseline;" id="TBL-2-2-"><td  style="white-space:nowrap; text-align:left;" id="TBL-2-2-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-2-2"  
class="td11">10               </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-2-3"  
class="td11">100        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-2-4"  
class="td11">1000               </td></tr><tr  
 style="vertical-align:baseline;" id="TBL-2-3-"><td  style="white-space:nowrap; text-align:left;" id="TBL-2-3-1"  
class="td11">0.40 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-3-2"  
class="td11">10 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-3-3"  
class="td11">1000 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-3-4"  
class="td11">1000</td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-4-"><td  style="white-space:nowrap; text-align:left;" id="TBL-2-4-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-4-2"  
class="td11">10               </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-4-3"  
class="td11">10000     </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-4-4"  
class="td11">1000               </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-5-"><td  style="white-space:nowrap; text-align:left;" id="TBL-2-5-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-5-2"  
class="td11">10               </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-5-3"  
class="td11">100000    </td><td  style="white-space:nowrap; text-align:left;" id="TBL-2-5-4"  
class="td11">1000               </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-2-6-"><td  style="white-space:nowrap; text-align:left;" id="TBL-2-6-1"  
class="td11">         </td></tr></table></div>
<br /> <div class="caption" 
><span class="id">Table&#x00A0;3.2: </span><span  
class="content">Determining size scale effects.</span></div><!--tex4ht:label?: x1-130032 -->
</div>

       </div><hr class="endfloat" />
       </div>
<!--l. 48--><p class="indent" >       Figure&#x00A0;<a 
href="#x1-130042">3.2<!--tex4ht:ref: fig:scale_num --></a> is a plot of the average polymerase rate of the population in each
environment against the simulation time. Based on these initial experiments, it was decided
that a maximum population size of 1000 with a genome length of 1000 could serve as an
adequate representative of the dynamics over the range of possible values. These values
were chosen because they also keep the size of the simulations reasonable with regards to
the amount of computational time required to run each simulation, since the run time of
the simulations scale with the population size and organisms with longer genomes require
more time-steps to achieve the same number of doublings as organisms with shorter
genomes.
<!--l. 50--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-130042"></a>


<!--l. 52--><p class="noindent" ><img 
src="scale_num.png" alt="PIC"  
width="433.62pt" height="261.17322pt" >
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;3.2:  </span><span  
class="content"><span 
class="cmbx-12">Effects  of  Environmental  Carrying  Capacity  on  Polymerase</span>
<span 
class="cmbx-12">Rate Evolution. </span>The average polymerase rate for the entire simulation population
is plotted against simulation time steps (each unit on the abscissa is equivalent to
100 simulation time steps). The organisms each had a genome length of 1000, and the
environments had a maximum capacity (<span 
class="cmmi-12">N</span>) of 100, 1000, 10000, or 100000 organisms.</span></div><!--tex4ht:label?: x1-130042 -->

<!--l. 55--><p class="indent" >       </div><hr class="endfigure">
<!--l. 57--><p class="indent" >       Finally, in order to validate the model and how reasonably it simulates the observed
growth dynamics of biological organisms, a set of experiments were carried out
starting with small populations of 10 organisms, as before, and following their
growth at different temperatures. The first two sets of experiments described
above were only carried out with forward polymerizing organisms. To be sure that
model organisms with reverse, 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032;</span>, polymerases had growth dynamics similar
to forward polymerizing organisms, these experiments were carried out using
both types of organisms. Table&#x00A0;<a 
href="#x1-130053">3.3<!--tex4ht:ref: tab:temp_incr --></a> summarizes the parameters used for these
experiments.
       <div class="table">

<!--l. 60--><p class="indent" >       <a 
 id="x1-130053"></a><hr class="float"><div class="float" 
>

<div class="center" 
>
<!--l. 60--><p class="noindent" >
<div class="tabular"> <table id="TBL-3" class="tabular" 
cellspacing="0" cellpadding="0" rules="groups" 
><colgroup id="TBL-3-1g"><col 
id="TBL-3-1"></colgroup><colgroup id="TBL-3-2g"><col 
id="TBL-3-2"></colgroup><colgroup id="TBL-3-3g"><col 
id="TBL-3-3"></colgroup><colgroup id="TBL-3-4g"><col 
id="TBL-3-4"></colgroup><tr  
 style="vertical-align:baseline;" id="TBL-3-1-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-1-1"  
class="td11">Temp. (<span 
class="cmmi-12">t</span>)</td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-1-2"  
class="td11">Max Pop.</td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-1-3"  
class="td11">Genome Length</td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-1-4"  
class="td11">Directionality</td>
</tr><tr 
class="hline"><td><hr></td><td><hr></td><td><hr></td><td><hr></td></tr><tr  
 style="vertical-align:baseline;" id="TBL-3-2-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-2-1"  
class="td11">0.10        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-2-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-2-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-2-4"  
class="td11">           </td></tr><tr  
 style="vertical-align:baseline;" id="TBL-3-3-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-3-1"  
class="td11">0.30 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-3-2"  
class="td11"> </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-3-3"  
class="td11"> </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-3-4"  
class="td11">forward</td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-4-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-4-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-4-2"  
class="td11">1000       </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-4-3"  
class="td11">1000               </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-4-4"  
class="td11">           </td></tr><tr  
 style="vertical-align:baseline;" id="TBL-3-5-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-5-1"  
class="td11">0.60 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-5-2"  
class="td11"> </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-5-3"  
class="td11"> </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-5-4"  
class="td11"></td>
</tr><tr 
class="hline"><td><hr></td><td><hr></td><td><hr></td><td><hr></td></tr><tr  
 style="vertical-align:baseline;" id="TBL-3-6-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-6-1"  
class="td11">0.10        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-6-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-6-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-6-4"  
class="td11">           </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-7-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-7-1"  
class="td11">0.30        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-7-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-7-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-7-4"  
class="td11">reverse         </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-8-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-8-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-8-2"  
class="td11">1000       </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-8-3"  
class="td11">1000               </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-8-4"  
class="td11">           </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-9-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-9-1"  
class="td11">0.60        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-9-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-9-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:left;" id="TBL-3-9-4"  
class="td11">           </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-3-10-"><td  style="white-space:nowrap; text-align:left;" id="TBL-3-10-1"  
class="td11">         </td></tr></table></div>
<br /> <div class="caption" 
><span class="id">Table&#x00A0;3.3: </span><span  
class="content">Growth dynamics at various temperatures.</span></div><!--tex4ht:label?: x1-130053 -->
</div>

       </div><hr class="endfloat" />
       </div>
<!--l. 79--><p class="indent" >       The results of these simulations are presented in figure&#x00A0;<a 
href="#x1-130063">3.3<!--tex4ht:ref: fig:temp_incr_mut --></a>. Simulations were carried
out for each type of polymerase individually so that only the dynamics of the polymerase
would influence growth. The simulations were also run disallowing mutations. Under this
condition there can be no change in polymerase rate for an individual and its daughters
from one generation to the next. These results are presented in figure&#x00A0;<a 
href="#x1-130074">3.4<!--tex4ht:ref: fig:temp_incr_nomut --></a>. We monitored
both the population rate and the average polymerase rate for all the organisms in the
environment.
<!--l. 81--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-130063"></a>

<div class="center" 
>
<!--l. 82--><p class="noindent" >

<!--l. 83--><p class="noindent" ><img 
src="temp_incr_mut.png" alt="PIC"  
width="411.93767pt" height="480.7122pt" ></div>
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;3.3: </span><span  
class="content"><span 
class="cmbx-12">Exponential growth at various temperatures in the absence of</span>
<span 
class="cmbx-12">competition, with mutations. </span>The model system was seeded with environments,
at simulation temperatures of 0.10, 0.30, 0.40, or 0.60, containing 10 organisms with
a 5.5 average polymerase rate. <span 
class="cmbx-12">A</span>. Population size of model organisms as a function
of simulation time. <span 
class="cmbx-12">B</span>. Evolution of the average polymerase rate for the organisms in
each environment as a function of simulation time. In each case, solid lines are used to
indicate environments with forward polymerizing organisms and dashed lines are for
reverse polymerizing organisms. Different temperatures are indicated with different
data markers as indicated in the figure legend, and are expressed in units of <img 
src="steventhesis_num10x.png" alt="&#x0394;RE-"  class="frac" align="middle">. </span></div><!--tex4ht:label?: x1-130063 -->

<!--l. 89--><p class="indent" >       </div><hr class="endfigure">
<!--l. 91--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-130074"></a>

<div class="center" 
>
<!--l. 92--><p class="noindent" >

<!--l. 93--><p class="noindent" ><img 
src="temp_incr_nomut.png" alt="PIC"  
width="411.93767pt" height="480.7122pt" ></div>
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;3.4: </span><span  
class="content"><span 
class="cmbx-12">Exponential growth at various temperatures in the absence</span>
<span 
class="cmbx-12">of  competition,  with  no  mutations.  </span>The  model  system  was  seeded  with
environments, at simulation temperatures of 0.10, 0.30, 0.40, or 0.60, containing 10
organisms with a 5.5 average polymerase rate. <span 
class="cmbx-12">A</span>. Population size of model organisms
as a function of simulation time. <span 
class="cmbx-12">B</span>. Evolution of the average polymerase rate for the
organisms in each environment as a function of simulation time. In each case, solid
lines are used to indicate environments with forward polymerizing organisms and
dashed lines are for reverse polymerizing organisms. For every simulation mutations
were disallowed. Different temperatures are indicated with different data markers as
indicated in the figure legend, and are expressed in units of <img 
src="steventhesis_num11x.png" alt="&#x0394;ER-"  class="frac" align="middle">. </span></div><!--tex4ht:label?: x1-130074 -->

<!--l. 99--><p class="indent" >       </div><hr class="endfigure">
       <h5 class="likesubsubsectionHead"><a 
 id="x1-140003"></a>Evolution During Exponential Growth</h5>
<!--l. 103--><p class="noindent" >In each of the experiments performed up to this point, forward or reverse organisms were
allowed to grow in isolation. In order to gain some insight into the dynamics of
polymerase evolution, it is necessary to set these two classes of model organisms in
competition with each other. When considering the competition of organisms, there are
two domains which are interesting to probe. The first is the competition of the
organisms as they explore a new ecological niche. That is, the way in which organisms
compete during phases of exponential growth. The second is the competition
that occurs when an environment is already at its carrying capacity. It would be
expected that an evolutionary strategy which results in the most rapid growth
should dominate during exponential growth. It is also conceivable that such a
strategy might not represent the most efficient use of resources available and
therefore might ultimately loose out to a different strategy when growth is limited by
resources.
<!--l. 105--><p class="indent" >       We investigated how organisms with forward or reverse polymerases would behave
when competing with each other for resources during an exponential growth phase. To do
this, we seeded environments with 100 organisms, where 50 of the organisms contained
forward polymerases and 50 reverse. For each type, there were 5 organisms with each
possible polymerase rates for an average starting polymerase rate of 5.5. Table&#x00A0;<a 
href="#x1-140014">3.4<!--tex4ht:ref: tab:growth_compet --></a>
summarizes the parameters used for these experiments.
       <div class="table">

<!--l. 108--><p class="indent" >       <a 
 id="x1-140014"></a><hr class="float"><div class="float" 
>

<div class="center" 
>
<!--l. 108--><p class="noindent" >
<div class="tabular"> <table id="TBL-4" class="tabular" 
cellspacing="0" cellpadding="0" rules="groups" 
><colgroup id="TBL-4-1g"><col 
id="TBL-4-1"></colgroup><colgroup id="TBL-4-2g"><col 
id="TBL-4-2"></colgroup><colgroup id="TBL-4-3g"><col 
id="TBL-4-3"></colgroup><colgroup id="TBL-4-4g"><col 
id="TBL-4-4"></colgroup><tr  
 style="vertical-align:baseline;" id="TBL-4-1-"><td  style="white-space:nowrap; text-align:left;" id="TBL-4-1-1"  
class="td11">Temp. (<span 
class="cmmi-12">t</span>)</td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-1-2"  
class="td11">Max Pop.</td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-1-3"  
class="td11">Genome Length</td><td  style="white-space:nowrap; text-align:center;" id="TBL-4-1-4"  
class="td11">Seed organisms</td>
</tr><tr 
class="hline"><td><hr></td><td><hr></td><td><hr></td><td><hr></td></tr><tr  
 style="vertical-align:baseline;" id="TBL-4-2-"><td  style="white-space:nowrap; text-align:left;" id="TBL-4-2-1"  
class="td11">0.10        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-2-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-2-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-4-2-4"  
class="td11">  50 forward   </td></tr><tr  
 style="vertical-align:baseline;" id="TBL-4-3-"><td  style="white-space:nowrap; text-align:left;" id="TBL-4-3-1"  
class="td11">0.30 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-3-2"  
class="td11"> </td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-3-3"  
class="td11"> </td><td  style="white-space:nowrap; text-align:center;" id="TBL-4-3-4"  
class="td11"> and</td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-4-4-"><td  style="white-space:nowrap; text-align:left;" id="TBL-4-4-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-4-2"  
class="td11">1000       </td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-4-3"  
class="td11">1000               </td><td  style="white-space:nowrap; text-align:center;" id="TBL-4-4-4"  
class="td11">  50 reverse    </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-4-5-"><td  style="white-space:nowrap; text-align:left;" id="TBL-4-5-1"  
class="td11">0.60        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-5-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-4-5-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-4-5-4"  
class="td11">             </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-4-6-"><td  style="white-space:nowrap; text-align:left;" id="TBL-4-6-1"  
class="td11">         </td></tr></table></div>
<br /> <div class="caption" 
><span class="id">Table&#x00A0;3.4: </span><span  
class="content">Competitive growth at various temperatures.</span></div><!--tex4ht:label?: x1-140014 -->
</div>

       </div><hr class="endfloat" />
       </div>
<!--l. 122--><p class="indent" >       The results of these simulations are presented in figure&#x00A0;<a 
href="#x1-140025">3.5<!--tex4ht:ref: fig:growth_compet_mut --></a>. Simulations were again
replicated disallowing mutations. These results are presented in figure&#x00A0;<a 
href="#x1-140036">3.6<!--tex4ht:ref: fig:growth_compet_nomut --></a>. In each
environment, we tracked the subpopulations of organisms with forward and reverse
polymerases separately so that we could follow the population changes and polymerase rate
evolution for each condition in the mixed case independently.
<!--l. 124--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-140025"></a>

<div class="center" 
>
<!--l. 125--><p class="noindent" >

<!--l. 126--><p class="noindent" ><img 
src="growth_compet_mut.png" alt="PIC"  
width="411.93767pt" height="480.7122pt" ></div>
<br />                <div class="caption" 
><span class="id">Figure&#x00A0;3.5:                     </span><span  
class="content"><span 
class="cmbx-12">Competition                     during</span>
<span 
class="cmbx-12">exponential growth at various temperatures, with mutations. </span>The model
system was seeded with environments, at simulation temperatures of 0.10, 0.30, 0.40,
or 0.60, containing 100 organisms, 50 each with forward and reverse polymerases, with
a 5.5 average polymerase rate. <span 
class="cmbx-12">A</span>. Population size of model organisms as a function
of simulation time. <span 
class="cmbx-12">B</span>. Evolution of the average polymerase rate for the organisms in
each environment as a function of simulation time. In each case, solid lines are used to
indicate environments with forward polymerizing organisms and dashed lines are for
reverse polymerizing organisms. Different temperatures are indicated with different
data markers as indicated in the figure legend, and are expressed in units of <img 
src="steventhesis_num12x.png" alt="&#x0394;RE-"  class="frac" align="middle">. </span></div><!--tex4ht:label?: x1-140025 -->

<!--l. 132--><p class="indent" >       </div><hr class="endfigure">
<!--l. 134--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-140036"></a>

<div class="center" 
>
<!--l. 135--><p class="noindent" >

<!--l. 136--><p class="noindent" ><img 
src="growth_compet_nomut.png" alt="PIC"  
width="411.93767pt" height="480.7122pt" ></div>
<br />                <div class="caption" 
><span class="id">Figure&#x00A0;3.6:                     </span><span  
class="content"><span 
class="cmbx-12">Competition                     during</span>
<span 
class="cmbx-12">exponential growth at various temperatures, with no mutations. </span>The model
system was seeded with environments, at simulation temperatures of 0.10, 0.30, 0.40,
or 0.60, containing 100 organisms, 50 each with forward and reverse polymerases, with
a 5.5 average polymerase rate. <span 
class="cmbx-12">A</span>. Population size of model organisms as a function
of simulation time. <span 
class="cmbx-12">B</span>. Evolution of the average polymerase rate for the organisms in
each environment as a function of simulation time. In each case, solid lines are used
to indicate environments with forward polymerizing organisms and dashed lines are
for reverse polymerizing organisms. For every simulation mutations were disallowed.
Different temperatures are indicated with different data markers as indicated in the
figure legend, and are expressed in units of <img 
src="steventhesis_num13x.png" alt="&#x0394;ER-"  class="frac" align="middle">. </span></div><!--tex4ht:label?: x1-140036 -->

<!--l. 142--><p class="indent" >       </div><hr class="endfigure">
       <h5 class="likesubsubsectionHead"><a 
 id="x1-150003"></a>Competition in a Full Environment</h5>
<!--l. 145--><p class="noindent" >To further investigate how organisms with forward polymerases fared in competition with
organisms with reverse polymerases, specifically to see if an equilibrium was possible
between the two varieties, we performed a number of simulations starting with an
environment already at its maximum capacity. This was done by seeding each
environment with 500 organisms containing forward polymerases and 500 containing
reverse. For each variety, there were 50 organisms seeded with each of the 10
possible polymerase rates. Table&#x00A0;<a 
href="#x1-150015">3.5<!--tex4ht:ref: tab:strict --></a> summarizes the parameters used for these
experiments.
       <div class="table">

<!--l. 148--><p class="indent" >       <a 
 id="x1-150015"></a><hr class="float"><div class="float" 
>

<div class="center" 
>
<!--l. 148--><p class="noindent" >
<div class="tabular"> <table id="TBL-5" class="tabular" 
cellspacing="0" cellpadding="0" rules="groups" 
><colgroup id="TBL-5-1g"><col 
id="TBL-5-1"></colgroup><colgroup id="TBL-5-2g"><col 
id="TBL-5-2"></colgroup><colgroup id="TBL-5-3g"><col 
id="TBL-5-3"></colgroup><colgroup id="TBL-5-4g"><col 
id="TBL-5-4"></colgroup><tr  
 style="vertical-align:baseline;" id="TBL-5-1-"><td  style="white-space:nowrap; text-align:left;" id="TBL-5-1-1"  
class="td11">Temp. (<span 
class="cmmi-12">t</span>)</td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-1-2"  
class="td11">Max Pop.</td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-1-3"  
class="td11">Genome Length</td><td  style="white-space:nowrap; text-align:center;" id="TBL-5-1-4"  
class="td11">Seed organisms</td>
</tr><tr 
class="hline"><td><hr></td><td><hr></td><td><hr></td><td><hr></td></tr><tr  
 style="vertical-align:baseline;" id="TBL-5-2-"><td  style="white-space:nowrap; text-align:left;" id="TBL-5-2-1"  
class="td11">0.10        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-2-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-2-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-5-2-4"  
class="td11"> 500 forward  </td></tr><tr  
 style="vertical-align:baseline;" id="TBL-5-3-"><td  style="white-space:nowrap; text-align:left;" id="TBL-5-3-1"  
class="td11">0.30 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-3-2"  
class="td11"> </td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-3-3"  
class="td11"> </td><td  style="white-space:nowrap; text-align:center;" id="TBL-5-3-4"  
class="td11"> and</td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-5-4-"><td  style="white-space:nowrap; text-align:left;" id="TBL-5-4-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-4-2"  
class="td11">1000       </td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-4-3"  
class="td11">1000               </td><td  style="white-space:nowrap; text-align:center;" id="TBL-5-4-4"  
class="td11">  500 reverse   </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-5-5-"><td  style="white-space:nowrap; text-align:left;" id="TBL-5-5-1"  
class="td11">0.60        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-5-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-5-5-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-5-5-4"  
class="td11">             </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-5-6-"><td  style="white-space:nowrap; text-align:left;" id="TBL-5-6-1"  
class="td11">         </td></tr></table></div>
<br /> <div class="caption" 
><span class="id">Table&#x00A0;3.5: </span><span  
class="content">Competitive growth in a full environment at various temperatures.</span></div><!--tex4ht:label?: x1-150015 -->
</div>

       </div><hr class="endfloat" />
       </div>
<!--l. 162--><p class="indent" >       The results of these simulations are presented in figure&#x00A0;<a 
href="#x1-150027">3.7<!--tex4ht:ref: fig:strict_compet_mut --></a>. As before, these
conditions were repeated with mutations disallowed, and those results are presented in
figure&#x00A0;<a 
href="#x1-150038">3.8<!--tex4ht:ref: fig:strict_compet_nomut --></a>.
<!--l. 164--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-150027"></a>

<div class="center" 
>
<!--l. 165--><p class="noindent" >

<!--l. 166--><p class="noindent" ><img 
src="strict_compet_mut.png" alt="PIC"  
width="411.93767pt" height="480.7122pt" ></div>
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;3.7: </span><span  
class="content"><span 
class="cmbx-12">Competition in an environment at maximum capacity, with</span>
<span 
class="cmbx-12">mutations.  </span>Environments,  at  simulation  temperatures  of  0.10,  0.30,  0.40,  or
0.60,  were  seeded  with  500  organisms  containing  forward  polymerases  and  500
containing  reverse,  both  with  a  5.5  average  polymerase  rate.  <span 
class="cmbx-12">A</span>.  Population  size
of model organisms as a function of simulation time. <span 
class="cmbx-12">B</span>. Evolution of the average
polymerase rate for the organisms in each environment as a function of simulation
time.  In  each  case,  solid  lines  are  used  to  indicate  environments  with  forward
polymerizing  organisms  and  dashed  lines  are  for  reverse  polymerizing  organisms.
Different temperatures are indicated with different data markers as indicated in the
figure legend, and are expressed in units of <img 
src="steventhesis_num14x.png" alt="&#x0394;ER-"  class="frac" align="middle">. </span></div><!--tex4ht:label?: x1-150027 -->

<!--l. 172--><p class="indent" >       </div><hr class="endfigure">
<!--l. 174--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-150038"></a>

<div class="center" 
>
<!--l. 175--><p class="noindent" >

<!--l. 176--><p class="noindent" ><img 
src="strict_compet_nomut.png" alt="PIC"  
width="411.93767pt" height="480.7122pt" ></div>
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;3.8: </span><span  
class="content"><span 
class="cmbx-12">Competition in an environment at maximum capacity, with</span>
<span 
class="cmbx-12">no mutations. </span>Environments, at simulation temperatures of 0.10, 0.30, 0.40, or
0.60,  were  seeded  with  500  organisms  containing  forward  polymerases  and  500
containing  reverse,  both  with  a  5.5  average  polymerase  rate.  <span 
class="cmbx-12">A</span>.  Population  size
of model organisms as a function of simulation time. <span 
class="cmbx-12">B</span>. Evolution of the average
polymerase rate for the organisms in each environment as a function of simulation
time.  In  each  case,  solid  lines  are  used  to  indicate  environments  with  forward
polymerizing organisms and dashed lines are for reverse polymerizing organisms. For
every simulation mutations were disallowed. Different temperatures are indicated with
different data markers as indicated in the figure legend, and are expressed in units of
<img 
src="steventhesis_num15x.png" alt="&#x0394;ER-"  class="frac" align="middle">. </span></div><!--tex4ht:label?: x1-150038 -->

<!--l. 182--><p class="indent" >       </div><hr class="endfigure">
<!--l. 184--><p class="indent" >       We noted that the results of this last experiment seemed to indicate that, when
mutations were allowed, their might be a temperature regime in which there is a transition
between an equilibrium of organisms containing forward and reverse polymerases and a
complete dominance by organisms containing the forward polymerase. To get a more
detailed picture of this transition temperature, we repeated the simulations with full
environments at temperatures from 0.10 to 0.60 in 0.05 increments. Table&#x00A0;<a 
href="#x1-150046">3.6<!--tex4ht:ref: tab:strict_log --></a> summarizes
the parameters used for these experiments, and figure&#x00A0;<a 
href="#x1-150059">3.9<!--tex4ht:ref: fig:strict_compet_log --></a> shows the results of these
simulations.
       <div class="table">

<!--l. 187--><p class="indent" >       <a 
 id="x1-150046"></a><hr class="float"><div class="float" 
>

<div class="center" 
>
<!--l. 187--><p class="noindent" >
<div class="tabular"> <table id="TBL-6" class="tabular" 
cellspacing="0" cellpadding="0" rules="groups" 
><colgroup id="TBL-6-1g"><col 
id="TBL-6-1"></colgroup><colgroup id="TBL-6-2g"><col 
id="TBL-6-2"></colgroup><colgroup id="TBL-6-3g"><col 
id="TBL-6-3"></colgroup><colgroup id="TBL-6-4g"><col 
id="TBL-6-4"></colgroup><tr  
 style="vertical-align:baseline;" id="TBL-6-1-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-1-1"  
class="td11">Temp. (<span 
class="cmmi-12">t</span>)</td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-1-2"  
class="td11">Max Pop.</td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-1-3"  
class="td11">Genome Length</td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-1-4"  
class="td11">Seed organisms</td>
</tr><tr 
class="hline"><td><hr></td><td><hr></td><td><hr></td><td><hr></td></tr><tr  
 style="vertical-align:baseline;" id="TBL-6-2-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-2-1"  
class="td11">0.10        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-2-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-2-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-2-4"  
class="td11">             </td></tr><tr  
 style="vertical-align:baseline;" id="TBL-6-3-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-3-1"  
class="td11">0.15 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-3-2"  
class="td11"> </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-3-3"  
class="td11"> </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-3-4"  
class="td11"></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-4-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-4-1"  
class="td11">0.20        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-4-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-4-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-4-4"  
class="td11"> 500 forward  </td></tr><tr  
 style="vertical-align:baseline;" id="TBL-6-5-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-5-1"  
class="td11">0.25 </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-5-2"  
class="td11"> </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-5-3"  
class="td11"> </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-5-4"  
class="td11"></td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-6-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-6-1"  
class="td11">0.30        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-6-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-6-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-6-4"  
class="td11">             </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-7-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-7-1"  
class="td11">0.35        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-7-2"  
class="td11">1000       </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-7-3"  
class="td11">1000               </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-7-4"  
class="td11">     and        </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-8-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-8-1"  
class="td11">0.40        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-8-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-8-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-8-4"  
class="td11">             </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-9-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-9-1"  
class="td11">0.45        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-9-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-9-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-9-4"  
class="td11">             </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-10-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-10-1"  
class="td11">0.50        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-10-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-10-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-10-4"  
class="td11">  500 reverse   </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-11-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-11-1"  
class="td11">0.55        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-11-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-11-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-11-4"  
class="td11">             </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-12-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-12-1"  
class="td11">0.60        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-12-2"  
class="td11">        </td><td  style="white-space:nowrap; text-align:left;" id="TBL-6-12-3"  
class="td11">             </td><td  style="white-space:nowrap; text-align:center;" id="TBL-6-12-4"  
class="td11">             </td>
</tr><tr  
 style="vertical-align:baseline;" id="TBL-6-13-"><td  style="white-space:nowrap; text-align:left;" id="TBL-6-13-1"  
class="td11">         </td></tr></table></div>
<br /> <div class="caption" 
><span class="id">Table&#x00A0;3.6:  </span><span  
class="content">Competitive  growth  in  a  full  environment  at  many  finely  divided
temperatures.</span></div><!--tex4ht:label?: x1-150046 -->
</div>

       </div><hr class="endfloat" />
       </div>
<!--l. 208--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-150059"></a>

<div class="center" 
>
<!--l. 209--><p class="noindent" >

<!--l. 210--><p class="noindent" ><img 
src="strict_compet_log.png" alt="PIC"  
width="411.93767pt" height="240.34875pt" ></div>
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;3.9: </span><span  
class="content"><span 
class="cmbx-12">Competition in an environment at maximum capacity at various</span>
<span 
class="cmbx-12">temperatures. </span>Simulations were carried out with environments at temperatures
ranging from 0.10 to 0.60 in 0.05 increments. In each simulation, the environment
was seeded at full capacity with 500 organisms containing forward polymerases and
500 containing reverse. For each variety there were 50 organisms with each of the
possible polymerase rates, giving an average rate of 5.5 In <span 
class="cmbx-12">A </span>and <span 
class="cmbx-12">B </span>the natural log of
the population of organisms containing reverse polymerases is plotted as a function
of simulation time. <span 
class="cmbx-12">A </span>is the data from simulations where mutation was allowed, and
<span 
class="cmbx-12">B </span>is from simulations where mutations were not permitted. A plot of the slopes of a
least squares regression line for the data in each simulations is plotted as a function of
simulation temperature. Data from simulations with mutation is plotted as the solid
line with data from the no mutation simulations plotted as a dashed line. </span></div><!--tex4ht:label?: x1-150059 -->

<!--l. 216--><p class="indent" >       </div><hr class="endfigure">



       <h2 class="chapterHead"><span class="titlemark">Chapter&#x00A0;4</span><br /><a 
 id="x1-160004"></a>Analysis and Discussion of Model Results</h2>
<!--l. 3--><p class="noindent" >In the study of evolutionary dynamics there exists the concept of a stable equilibrium. The
organisms in such a situation may not all represent the &#8220;best fit&#8221; mechanism for coping
with a particular ecological niche, but at the same time none are sufficiently better off than
the others that they will dominate the population. If organisms polymerizing nucleotides in
a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>direction represented a stable equilibrium with organisms polymerizing in a
5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>direction, then it would be expected that organisms exhibiting such a trait should
be found in nature. On the other hand, if 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerization is a loosing
strategy when in competition with 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerization, the lack of modern
3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerizing organisms would be an expected consequence of Darwinian
evolution.
<!--l. 5--><p class="indent" >       When modeling the competition between organisms with a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase and
organisms with a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase, this is an important aspect of modern evolutionary
theory to keep in mind. It is not sufficient to show that polymerizing nucleotides in a
5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>direction is better fit than, or produces organisms which reproduce faster than
polymerizing nucleotides in a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>direction. Rather, to discriminate between the
possibility that the modern ubiquity of 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerases is the result of a founder effect
versus being a consequence of natural selection, we need to show that 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerizing
organisms cannot exist in a stable equilibrium. It is with this in mind that the model
presented here was constructed.
<!--l. 7--><p class="indent" >       In constructing this model, a number of explicit and implicit assumptions were
made. The first, and perhaps largest, assumption is that polymerizing nucleotides to create
new nucleic acid polymers by adding nucleotide triphosphates to the 5<span 
class="cmsy-10x-x-120">&#x2032; </span>end of the growing
chain is physically possible. As there are no known polymerase enzymes which carry out
the polymerization reaction in a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>direction, it is impossible to say with
certainty that such polymerases might exist. However, based on the nature of

the active site chemistry of nucleotide polymerases, there is also no reason to
believe that such 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerases are not possible, beyond the fact that such
polymerases do not currently exist. For this reason, we are comfortable with this
assumption.
<!--l. 9--><p class="indent" >       The second implicit assumption made in constructing this model is that the only
viable building blocks for nucleic acid polymers are 5<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate nucleotides. If
nucleotides phosphorylated on their 3<span 
class="cmsy-10x-x-120">&#x2032; </span>oxygen could be used by a hypothetical 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032;</span>
polymerase, then the chemistry of such a polymerase would not, qualitatively, be
any different than that of a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase working with 5<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate
nucleotides. That is, in the case that a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase could work with a
3<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate nucleotide, then the activated triphosphate moiety would reside on the
incoming nucleotide, instead of on the growing chain, in the same way that the
triphosphate moiety is found on the incoming 5<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate nucleotide used by 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>
polymerases.
<!--l. 11--><p class="indent" >       We can discount the possibility of 3<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate nucleotides as raw material for a
polymerase reaction for a number of reasons. If, in keeping with current scientific
consensus, ribonucleic acids represent the more ancient form of nucleotides and the
2<span 
class="cmsy-10x-x-120">&#x2032;</span>-deoxyribonucleic acids are a more recent invention of biology, then a potential
3<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate nucleotide would have its triphosphate group attached on a carbon adjacent
to a free hydroxyl. The free 2<span 
class="cmsy-10x-x-120">&#x2032;</span>-hydroxyl group of ribonucleic acids is known to
polymerize hydrolysis reactions in a number of ribozymes, so it would not be
unreasonable to assume that the 2<span 
class="cmsy-10x-x-120">&#x2032;</span>-hydroxyl would also catalyze a hydrolysis of the
triphosphate group on a 3<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate nucleotide. This would serve to significantly
restrict the available primordial pool of 3<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate nucleotides as compared to
5<span 
class="cmsy-10x-x-120">&#x2032;</span>-triphosphate nucleotides, whose adjacent carbon&#8217;s oxygen would normally be
incorporated in the formation of the furanose ring. Indeed, it has been shown that

ribose-3<span 
class="cmsy-10x-x-120">&#x2032;</span>-phosphate decomposes more readily under mild acid conditions than
ribose-5<span 
class="cmsy-10x-x-120">&#x2032;</span>-phosphate.
<!--l. 13--><p class="indent" >       The remaining assumptions made are the various explicit assumptions required for
construction of the model itself as described in chapter&#x00A0;<a 
href="#x1-70002">2<!--tex4ht:ref: cha:design_of_the_polymerase_evolution_model --></a>. Of these, the most important is
the assumption that the dynamics of spontaneous nucleotide hydrolysis occur on a
time-scale commiserate with the time-scale of nucleotide incorporation and the
assumption of a geometry-based discrimination mechanism for incorrect nucleotide
incorporation by the polymerase. We can surmise that the time-scale for spontaneous
nucleotide hydrolysis cannot be significantly smaller than the time-scale for nucleotide
incorporation, for if it were then nucleotide polymerization would, in effect, be impossible.
If the time-scale were much larger, this would be the equivalent of removing a
constant factor from the calculation for the probability that nucleotide inactivation
occurs during polymerization. The effect of such an adjustment would be largely
quantitative, raising the simulation temperature at which the effects reported would be
observed.
<!--l. 15--><p class="indent" >       A note should be made on the significance of the simulation temperature used in the
model. If we take the expression used in the model for the temperature dependence of the
equilibrium between the polymerase-template-nucleotide complex for the correct
nucleotide, the free polymerase-template complex, and the polymerase-template-nucleotide
complex for the incorrect nucleotide: <span 
class="cmmi-12">e</span><sup><span 
class="cmsy-8">-</span><img 
src="steventhesis_num16x.png" alt="1
t"  class="frac" align="middle">
    </sup>, and compare it to the experimentally determined
value of 5<img 
src="steventhesis_num17x.png" alt="kmcoall"  class="frac" align="middle"> for the &#x0394;&#x0394;<span 
class="cmmi-12">G </span>of this system, we can calculate the real temperature
corresponding to each simulation temperature with the expression
       <center class="math-display" >
<img 
src="steventhesis_num18x.png" alt="      5kcal
T =  t-mol-
       R
" class="math-display" ></center>
<!--l. 17--><p class="nopar" > This gives us a real temperature of <span 
class="cmsy-10x-x-120">-</span>22<sup><span 
class="cmsy-8">&#x2218;</span></sup><span 
class="cmmi-12">C </span>for a simulation temperature of 0.10 and a real
temperature of 1200<sup><span 
class="cmsy-8">&#x2218;</span></sup><span 
class="cmmi-12">C </span>for a simulation temperature of 0.60. Obviously this range of
temperatures is unreasonably large for life, even when accounting for aqueous environments
with high salinity or at high pressures.
<!--l. 20--><p class="indent" >       As mentioned in the introduction, every effort was made when designing this model
to remain faithful to the realities of the biochemical processes that a nucleotide polymerase
must carry out. That said, a number of confounding factors limit the applicability of the
precise quantitative temperature values used in the simulations to understanding real
biological processes. It has already been mentioned that the need to reduce the probability
of spontaneous nucleotide hydrolysis by a constant factor to account for a difference in
time-scales would effect the absolute temperature values at which various simulation
outcomes are observed. The same sort of constant factor consideration would,
potentially, also need to be made to account for a difference between the precise effect
of geometry-based discrimination of real nucleotide polymerases and the model
values.
<!--l. 22--><p class="indent" >       The primary model values impacted by simulation temperature are those for the
equilibrium describing the incorporation of correct versus incorrect nucleotides during
nucleic acid synthesis, and the equilibrium between active and inactive nucleotides. The
chemical phenomena described by these values, however, are subject to more
influences than just temperature. For example, it is well documented that the
energy associated with a correct nucleotide base-pairing is heavily influenced
by the dielectric constant of the solvent and the concentration of nucleic acid
counter-ions. The equilibrium of nucleotide triphosphates would, likewise, be subject to

influence by effects of nucleotide concentration and pH. Thus, the value of simulation
temperature can be thought of as a stand in for a number of thermodynamic factors
that might influence the biochemical processes involved in the model processes.
Therefore, the primary utility of the simulation temperature is in revealing the
collected influence of thermodynamics on the system. Before we consider these
influences, let us look at the results of the initial experiments performed with the
model.
       <h5 class="likesubsubsectionHead"><a 
 id="x1-170004"></a>Effect of Temperature on Evolution</h5>
<!--l. 26--><p class="noindent" >In constructing the model presented here, we were attempting to evaluate the plausibility
of the hypothesis that all life contains nucleotide polymerases which proceed in a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032;</span>
direction due to an evolutionary advantage of this mechanism versus the reverse
3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>direction. In order to address this question while still remaining biologically
relevant, it was necessary to account for a pair of temperature dependent chemical
processes: spontaneous hydrolysis of nucleotide triphosphates and inclusion error
rate.
<!--l. 28--><p class="indent" >       First, to validate the kinetics of the model we carried out a number of simulations in
which each sort of organism, those containing the 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>(forward) polymerase and those
containing the 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>(reverse) polymerase, were allowed to grow from a small starting
population in the absence of competition (fig.&#x00A0;<a 
href="#x1-130063">3.3<!--tex4ht:ref: fig:temp_incr_mut --></a><span 
class="cmbx-12">A</span>). These experiments revealed that the
temperature of the environment had a gradual impact on the growth rate of the model
organisms, with the greatest inhibition on growth rate occurring at the highest
temperatures investigated. At these higher temperatures, the effect on growth appeared to
be greater on organisms containing the reverse polymerase than on those containing the
forward polymerase. This indicates that the disadvantages inherent in polymerizing
nucleic acids in a 3<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>5<span 
class="cmsy-10x-x-120">&#x2032; </span>direction become more exaggerated as the temperature

increases.
<!--l. 30--><p class="indent" >       Another way to analyze these experiments is to look at the way that the rate of the
nucleotide polymerases evolves at each temperature. Generally, we can state that tendency
toward a faster polymerase indicates that the predominant evolutionary pressure is on
reproduction rate, while a tendency toward slower polymerases indicates an increased
importance of reducing the mutation rate. This stems from the fact that a faster
polymerase will have a greater error rate. With this in mind, figure&#x00A0;<a 
href="#x1-130063">3.3<!--tex4ht:ref: fig:temp_incr_mut --></a><span 
class="cmbx-12">B </span>shows that, as the
temperature of the simulation increases, the average polymerase rate decreases
indicating the increasing importance of avoiding errors. This is true up to simulation
temperatures of 0.40, but at a simulation temperature of 0.60 the trend reverses
indicating a switch back toward reproduction rate as the primary evolutionary
pressure.
<!--l. 32--><p class="indent" >       This switch most likely represents a saturation of the effect that mutation can have
on the organisms. That is, in our model system an increasing mutation rate decreases the
fidelity with which each generation can pass along its traits (in this case just polymerase
rate) to subsequent generations. We expect that, at some high mutation rate, this fidelity
drops significantly enough that the polymerase rate inherited from one generation to the
next is not significantly distinguishable from a randomly assigned value. When the
temperature of the simulation gets high enough, the error rate due to the thermal
component will overwhelm the error rate due to polymerase rate, and reach this critical
value. At this point the selective pressure against faster polymerases resulting from the
need to preserve generation to generation fidelity will, in effect, vanish. In our system,
this leaves growth rate as the single remaining selective pressure, explaining the
reversal of the trend of polymerase rate evolution at high simulation temperatures.
Indeed, as can be seen in figure&#x00A0;<a 
href="#x1-170011">4.1<!--tex4ht:ref: fig:mut_rate --></a>, at simulation temperatures of 0.55 and 0.60,
the average change in polymerase rate from one generation to the next is nearly

5, the maximum that would be expected if polymerase rates were inherited at
random.
<!--l. 34--><p class="indent" >       <hr class="figure"><div class="figure" 
>

<a 
 id="x1-170011"></a><a 
 id="x1-21000A"></a>

<div class="center" 
>
<!--l. 35--><p class="noindent" >

<!--l. 36--><p class="noindent" ><img 
src="mutation_rate.png" alt="PIC"  
width="433.62pt" height="261.22282pt" ></div>
<br /> <div class="caption" 
><span class="id">Figure&#x00A0;4.1:  </span><span  
class="content"><span 
class="cmbx-12">Generational  change  in  polymerase  rate  as  a  function  of</span>
<span 
class="cmbx-12">temperature. </span>The organisms from the systems plotted in <a 
href="#x1-130063">3.3<!--tex4ht:ref: fig:temp_incr_mut --></a> were analyzed for
the difference between the rate of their polymerase and the rate of their parent&#8217;s
polymerase. This difference is plotted for the various different temperatures simulated.
The maximal difference we would expect to see (i.e. in the case that inheritance was
purely random) would be 5. </span></div><!--tex4ht:label?: x1-170011 -->

<!--l. 42--><p class="indent" >       </div><hr class="endfigure">
<!--l. 44--><p class="indent" >       One interesting result from this first experiment that was not completely
anticipated was that the polymerase rate of the forward and reverse polymerizing
organisms would be so similar over a wide range of simulation temperatures. We
expected that reverse polymerizing organisms, with the increased penalty for a
spontaneous hydrolysis event, would have a greater selective pressure to evolve a faster
polymerase holding all else constant. This seems to not be the case, though, as
the average polymerase rate of the forward and reverse polymerizing organisms
at all but the highest temperature are indistinguishable. The only conclusion
we can draw from this observation is that the selective pressure on mutation
rate is sufficiently rigid enough to make the spontaneous hydrolysis penalty a
negligible influence on the evolution of polymerase rate. This conclusion is further
backed up by evidence from the simulations with no mutation allowed, as described
below.
<!--l. 46--><p class="indent" >       The next set of experiments was intended to look at how forward and reverse
polymerizing organisms would fare in competition with each other at various temperatures.
We again started each simulation with a small seed population and allowed these
populations to grow in an exponential fashion. As we expected based on the similarity of
growth rates for the forward and reverse polymerizing organisms growing in isolation, both
forms grew rapidly until the environment was at capacity with approximately half
of the population belonging to each form. At this point, competition kicks in
and at simulation temperatures of 0.1 an equilibrium between the two forms is
established. At a simulation temperature of 0.3, it is unclear in the duration of the
simulation run whether an equilibrium might eventually be established or whether the
reverse polymerizing organisms would eventually be outcompeted. At the higher
simulation temperatures of 0.4 and 0.6, the reverse polymerizing organisms are

outcompeted by the forward polymerizing forms, however at the temperature
of 0.4 this is a true out-competition, where the reverse polymerizing form grew
to occupy nearly half the population before being eradicated. At a simulation
temperature of 0.6, the reverse polymerizing organisms are never able to establish
themselves.
<!--l. 48--><p class="indent" >       At the higher temperatures, the average polymerase rate of the reverse polymerizing
organisms begins to drop off as their population numbers, as seen in figure&#x00A0;<a 
href="#x1-140025">3.5<!--tex4ht:ref: fig:growth_compet_mut --></a><span 
class="cmbx-12">B</span>, begin to
drop off. This drop is not, however, indicative of a reduction in the fitness of the reverse
polymerizing organisms. Rather, as the reverse polymerizing organisms are gradually
outcompeted by the forward polymerizing organisms their number is reduced. Since the
only mechanism in the model for removing organisms from the population is the density
dependent random culling, which members of the reverse polymerizing organism
population will be removed is likewise random. Since the organisms with a faster
polymerase rate will reproduce more rapidly, they will be present in greater number,
and therefore will be more likely to be removed from the environment first. In
other words, the rate of the decrease in average polymerase rate of the reverse
polymerizing organisms can be seen as a proxy measure for the rate at which
the forward polymerizing organisms are preferentially occupying the available
capacity in the environment created when a reverse polymerizing organism is
culled.
<!--l. 50--><p class="indent" >       Looking at the evolution of the polymerase rates of the forward and reverse
polymerizing organisms, it is apparent that the selective pressure on polymerase rate is
greater when there is competition involved at low temperatures. That is, at a simulation
temperatures of 0.3 and 0.4, the steady-state polymerase rate under competition is not
appreciably different than under the no competition growth case, but at a simulation
temperature of 0.1 the steady-state polymerase rate under competition is elevated as

compared to the average polymerase rate for both forward and reverse polymerizing
organisms growing in isolation. From this we can conclude that rapid growth takes
on an increased importance when competition with an alternative strategy is
involved.
<!--l. 52--><p class="indent" >       Because the reverse polymerizing organisms were not able to grow significantly at a
temperature of 0.6 under competition, and because of the ambiguity of the eventual fate of
the reverse polymerizing organisms at a temperature of 0.3, we were curious what
might happen if both forms started off at a significantly larger population size.
Specifically, we were interested to see if forward polymerizing organisms represented
an evolutionarily stable strategy at these temperatures<span class="cite">[<a 
href="#XANowak:2006p857">10</a>, chap. 4]</span>. In order to
investigate this question, we carried out simulations starting with completely full
environments split evenly between forward and reverse polymerizing organisms at various
temperatures.
<!--l. 54--><p class="indent" >       The results from these simulations (fig.&#x00A0;<a 
href="#x1-150027">3.7<!--tex4ht:ref: fig:strict_compet_mut --></a>) indicate that the ultimate success of the
forward and reverse strategies in competition with each other when starting from half of an
environment at capacity is similar to that during exponential growth, though the dynamics
are interestingly different. In general, we can conclude that forward and reverse
polymerizing organisms will establish an equilibrium at lower temperatures and that
forward polymerizing organisms are dominant at higher temperatures. That is, at
sufficiently high enough temperatures, polymerizing nucleotides in a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>direction
appears to be an evolutionarily stable strategy.
<!--l. 56--><p class="indent" >       Also, the steady state polymerase rates reached at each temperature are the same as
for the exponential growth case above. The simulations starting with a full population
differ from those starting with exponentially growing populations in how the steady state
is reached. Arrival at the steady state polymerase rate occurs much earlier in
the case of exponential growth. This reflects the fact that in a mostly empty

environment it is easier for the best fit organisms to rapidly dominate a population. In a
population already at capacity, this rebalancing of traits can only occur through
attrition, which is a more gradual process. Interestingly, the rate at which the
steady state polymerase rate is achieved appears to impact the distribution of the
equilibrium population. At a simulation temperature of 0.1, the equilibrium population
in the exponential growth case is evenly divided between forward and reverse
polymerizing organisms, but in the full environment case the split is closer to <img 
src="steventhesis_num19x.png" alt="34"  class="frac" align="middle"> forward
polymerizing organisms and <img 
src="steventhesis_num20x.png" alt="14"  class="frac" align="middle"> reverse. We believe that this difference may point toward
the existence of different domains of competition. That is, it may be possible
that the fitness of the model organisms is impacted by the population relative
to the environment&#8217;s carrying capacity. Since there are two fitness components
encoded in the polymerases of our model organisms, namely proliferation rate
and mutation rate, what we may be observing is the existence of multiple fitness
equilibria between the forward and reverse polymerizing organisms. Future models
wherein these fitness components could be decoupled might help resolve this
possibility.
       <h5 class="likesubsubsectionHead"><a 
 id="x1-180004"></a>Effect of Mutation on Evolution</h5>
<!--l. 60--><p class="noindent" >To better understand the role that mutation and the introduction of random variation from
generation to generation might have on the results of our simulations, we modified the
model system to remove the possibility for mutation. With this modification, the average
polymerase rate can still evolve, but only through selection of individuals. Also, since the
only evolutionary pressure acting on the model organisms is reproduction rate, we would
expect that all individuals should converge on the maximum polymerase rate. Indeed, this
is precisely the result we observe (fig.&#x00A0;<a 
href="#x1-130074">3.4<!--tex4ht:ref: fig:temp_incr_nomut --></a><span 
class="cmbx-12">B</span>). With the selective pressure of mutation

removed, we can also directly observe the impact of the penalty on reverse polymerizing
organisms due to spontaneous hydrolysis of nucleotide triphosphates. In figure&#x00A0;<a 
href="#x1-130074">3.4<!--tex4ht:ref: fig:temp_incr_nomut --></a><span 
class="cmbx-12">B</span>, we
can see that at simulation temperatures of 0.4 and 0.6 the reverse polymerizing
organisms evolve to faster polymerase rates sooner than their forward polymerizing
counterparts.
<!--l. 62--><p class="indent" >       We also note that the exponential growth of both the forward and reverse
polymerizing organisms is essentially identical at all temperatures except for 0.6 (fig.
&#x00A0;<a 
href="#x1-130074">3.4<!--tex4ht:ref: fig:temp_incr_nomut --></a><span 
class="cmbx-12">A</span>). Reverse polymerizing organisms at a simulation temperature of 0.6 are the
only group to deviate. This is due to the excessive penalty paid by a reverse
polymerase at this high temperature. Moreover, the reverse polymerizing organisms at a
simulation temperature of 0.4 appear to grow with identical kinetics to the forward
polymerizing organisms at this temperature, showing that the penalty of a reverse
polymerizing strategy can be compensated for by evolving a faster polymerase (as seen in
fig.&#x00A0;<a 
href="#x1-130074">3.4<!--tex4ht:ref: fig:temp_incr_nomut --></a><span 
class="cmbx-12">B</span>).
<!--l. 64--><p class="indent" >       Following this initial result, we then repeated the prior experiments starting with
small populations and starting with a full environment, but with mutations disallowed
(figures&#x00A0;<a 
href="#x1-140036">3.6<!--tex4ht:ref: fig:growth_compet_nomut --></a> &amp;&#x00A0;<a 
href="#x1-150038">3.8<!--tex4ht:ref: fig:strict_compet_nomut --></a>, respectively). When we start with small populations we see, as before,
that polymerase rate rapidly evolves to its equilibrium value (fig.&#x00A0;<a 
href="#x1-140036">3.6<!--tex4ht:ref: fig:growth_compet_nomut --></a><span 
class="cmbx-12">B</span>). The difference
here is that without mutation this equilibrium value at all temperatures is the
maximum polymerase rate allowed. At the highest temperature of 0.6, the reverse
polymerizing organisms are unable to establish themselves the same as when
mutations were allowed, implying that this inability to compete with forward
polymerizing organisms during exponential growth is a consequence of the spontaneous
hydrolysis penalty and is not affected by the presence or absence of mutations in the
system.
<!--l. 66--><p class="indent" >       Paying attention to the way that the competition between forward and reverse

polymerizing organisms resolves itself at a simulation temperature of 0.4 when starting
with a small population (fig.&#x00A0;<a 
href="#x1-140036">3.6<!--tex4ht:ref: fig:growth_compet_nomut --></a><span 
class="cmbx-12">A</span>), its not entirely clear whether an equilibrium between
these forms can be established. If we look for the establishment of an equilibrium when
starting with a full environment at the same temperature (fig&#x00A0;<a 
href="#x1-150038">3.8<!--tex4ht:ref: fig:strict_compet_nomut --></a><span 
class="cmbx-12">A</span>), it does appear that
the forward polymerizing organisms are outcompeting the reverse polymerizing organisms,
but the rate at which this take over happens is definitely slower than when mutations are
allowed (fig.&#x00A0;<a 
href="#x1-150027">3.7<!--tex4ht:ref: fig:strict_compet_mut --></a><span 
class="cmbx-12">A</span>).
<!--l. 68--><p class="indent" >       This leaves a question of whether allowing mutations, which affect the generation to
generation inheritance of polymerase rate, allow for a more rapid resolution of the more
dominant strategy, or whether the addition of mutation to a reverse polymerase strategy
causes this to be a loosing strategy when it otherwise might not be. We noticed that the
decline in population of reverse polymerizing organisms when starting from a full
environment is roughly exponential, so we plotted the natural log of this number against
simulation time in figures&#x00A0;<a 
href="#x1-150059">3.9<!--tex4ht:ref: fig:strict_compet_log --></a>, <span 
class="cmbx-12">A </span>&amp; <span 
class="cmbx-12">B </span>for simulations where mutations were allowed in the
system or where they were prohibited, respectively. These plots represent the rate of
population change in each simulation. By calculating the least squares regressions of
these rates and plotting the slope of these regressions versus the temperature
(fig.&#x00A0;<a 
href="#x1-150059">3.9<!--tex4ht:ref: fig:strict_compet_log --></a><span 
class="cmbx-12">C</span>), we can see clearly that there is an inflection point for both the mutation
allowed and the mutation disallowed situations, but that it occurs at different
temperatures.
<!--l. 70--><p class="indent" >       When an equilibrium between forward and reverse polymerizing organisms is
established, we expect the rate of population change to not be significantly different from
zero. At low simulation temperatures, this is indeed the case. The inflection of the plots in
figure&#x00A0;<a 
href="#x1-150059">3.9<!--tex4ht:ref: fig:strict_compet_log --></a><span 
class="cmbx-12">C </span>represents the point at which forward and reverse polymerizing organisms go
from being able to establish an equilibrium to a situation where the forward polymerizing
organisms are an evolutionarily stable strategy, and the reverse polymerizing organisms will

eventually be eradicated from the environment. That this inflection occurs at different
temperatures with or without mutations allowed indicates that there is a range of
simulation temperatures where mutation does, in fact, make the difference between a
forward polymerizing strategy being evolutionarily stable versus merely being the dominant
strategy.
<!--l. 73--><p class="noindent" >
       <h5 class="likesubsubsectionHead"><a 
 id="x1-190004"></a>Founder Effect or Evolution?</h5>
<!--l. 74--><p class="noindent" >Finally, returning to our original question of whether the present day situation that finds
all organisms polymerizing nucleotide polymers in the same 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>direction is the result
of a founder event or a consequence of evolution toward a best fit trait, unfortunately
the evidence is equivocal. Here we have presented a model system that could
potentially explain how evolution might account for the rise of a single polymerase
strategy.
<!--l. 76--><p class="indent" >       At the same time, we have shown that this strategy can coexist with the alternative
under the right conditions. That is, at a high temperature it seems clear that polymerizing
nucleotides 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>is clearly favored, and would be selected for fairly rapidly. At
lower temperatures evolution leads to a predominance of one form, which would
make a founder event resulting in a 5<span 
class="cmsy-10x-x-120">&#x2032;&#x2192; </span>3<span 
class="cmsy-10x-x-120">&#x2032; </span>polymerase more likely than the
reverse, but evolution alone cannot explain the convergence to a single polymerase
strategy.
<!--l. 78--><p class="indent" >       Really, our inability to link simulation temperature to an actual environmental
temperature with any great confidence makes it so that we must leave it at this. We can
state that life originating in proximity to a deep sea hydrothermal vent would be more
likely to arrive at a single strategy through evolution than life originating in Darwin&#8217;s

&#8220;warm puddle&#8221;, though it is possible that both environments would be either above or
below the temperature at which evolution leads to a single polymerase form. While this is
the most we can conclude from the evidence presented here, combined with other
evidence that the machinery involved in DNA polymerization evolved at least twice
independently<span class="cite">[<a 
href="#XLeipe:1999p1040">8</a>]</span>, the weight of evidence favors the explanation that evolutionary
competition was the driving factor in determining the modern strategy of nucleotide
synthesis.
<!--l. 81--><p class="noindent" >
       <h5 class="likesubsubsectionHead"><a 
 id="x1-200004"></a>Significance of Results</h5>
<!--l. 82--><p class="noindent" >Though the evidence presented may not provide conclusive support for or against our
original hypothesis, the system constructed and the results obtained from it do present
many lessons and interesting leads for future investigations. This is the first system that we
are aware of that combines both modeling of the &#8220;evolutionary short-loop&#8221;, where mutation
alters the replication mechanism which controls the rate of mutation, and simplified
real-world thermodynamics of the biochemical processes involved in life. For this reason,
this is also the first time we can look at the direct impact that an environmental factor
such as temperature has on evolving systems. This is important for the development of
theoretical approaches to evolution which depend on interactions between organisms and
their environments.
<!--l. 84--><p class="indent" >       Finally, the model we constructed allowed us to probe the consequences of mutation
on evolution. The role of mutations in evolution has long been debated. On one hand,
mutations provide variations which are the fodder of natural selection. On the other hand,
mutations reduce the information content passed from generation to generation, thereby
reducing the efficiency of selection. Our results indicate that the role of mutations in

evolution may be complicated by the nature of the evolving system. At low simulation
temperatures, mutations appeared to have little to no effect on evolution and competition
between forms. In a middle range of simulation temperatures, mutations are the difference
between a strategy being able to successful coexist with another and that strategy
being a loosing strategy that is eventually eliminated. Finally, at the highest
simulation temperatures, the only difference introduced by mutations is a difference in
the rate at which the successful strategy is able to outcompete the alternative.
This insight into the role that mutations play will surely be invaluable for future
investigations.



<a 
 id="x1-20001r33"></a>
       <h2 class="appendixHead"><span class="titlemark">Appendix&#x00A0;A</span><br /><a 
 id="x1-21000A"></a>Source Code</h2> Following is the Ruby source code for the four main classes
and the executable Ruby script used to run simulations based on YAML input files.
<!--l. 7--><div class="lstinputlisting">
<br /> <div class="caption" 
><span class="id">: </span><span  
class="content">constants.rb</span></div><!--tex4ht:label?: x1-21000A --><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21001r1"></a><span 
class="cmr-8">1</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Copyright</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">c</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">2009</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Joshua</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Ballanco</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21002r2"></a><span 
class="cmr-8">2</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21003r3"></a><span 
class="cmr-8">3</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">constants</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">rb</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21004r4"></a><span 
class="cmr-8">4</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21005r5"></a><span 
class="cmr-8">5</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">This</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">file</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">contains</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">constants</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">used</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">elsewhere</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">evolver</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">project</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21006r6"></a><span 
class="cmr-8">6</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21007r7"></a><span 
class="cmr-8">7</span></span><span 
class="cmtt-8">MAX_TOL_MUT_RATE</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0.34</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21008r8"></a><span 
class="cmr-8">8</span></span><span 
class="cmtt-8">MAX_POLY_RATE</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">10</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21009r9"></a><span 
class="cmr-8">9</span></span><span 
class="cmtt-8">MIN_POLY_RATE</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span>
       </div>
<!--l. 9--><div class="lstinputlisting">
<br /> <div class="caption" 
><span class="id">: </span><span  
class="content">environment.rb</span></div><!--tex4ht:label?: x1-21000A --><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21010r1"></a><span 
class="cmr-8">1</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Copyright</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">c</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">2009</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Joshua</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Ballanco</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21011r2"></a><span 
class="cmr-8">2</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21012r3"></a><span 
class="cmr-8">3</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Environment</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21013r4"></a><span 
class="cmr-8">4</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21014r5"></a><span 
class="cmr-8">5</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Abstract</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">contains</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">entire</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simulation</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">It</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21015r6"></a><span 
class="cmr-8">6</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">primarily</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">responsible</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">tracking</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">all</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simulation</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21016r7"></a><span 
class="cmr-8">7</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">calculating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">density</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">dependent</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">death</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">probability</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">randomly</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">culling</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21017r8"></a><span 
class="cmr-8">8</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">according</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">probability</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">stepping</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">each</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">at</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">each</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21018r9"></a><span 
class="cmr-8">9</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">time</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">step</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simulation</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21019r10"></a><span 
class="cmr-8">10</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21020r11"></a><span 
class="cmr-8">11</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">GenomeForEnvironment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">struct</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">used</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">seed</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">starting</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21021r12"></a><span 
class="cmr-8">12</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21022r13"></a><span 
class="cmr-8">13</span></span><span 
class="cmtt-8">GenomeForSpecies</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Struct</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">(:</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">population_frequency</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21023r14"></a><span 
class="cmr-8">14</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21024r15"></a><span 
class="cmr-8">15</span></span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Environment</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21025r16"></a><span 
class="cmr-8">16</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">attr_reader</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21026r17"></a><span 
class="cmr-8">17</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21027r18"></a><span 
class="cmr-8">18</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">must</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">be</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initialized</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">with</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21028r19"></a><span 
class="cmr-8">19</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simulation</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">maximum</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">starting</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">populations</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">GenomeForSpecies</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21029r20"></a><span 
class="cmr-8">20</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">array</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">enumerating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genomes</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">be</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">use</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">creating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initial</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21030r21"></a><span 
class="cmr-8">21</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initialize</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21031r22"></a><span 
class="cmr-8">22</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">max_population</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21032r23"></a><span 
class="cmr-8">23</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">starting_population</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21033r24"></a><span 
class="cmr-8">24</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">genomes_for_environment</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21034r25"></a><span 
class="cmr-8">25</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21035r26"></a><span 
class="cmr-8">26</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Some</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simple</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">sanity</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">checks</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">passed</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">arguments</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21036r27"></a><span 
class="cmr-8">27</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">unless</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genomes_for_environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">inject</span><span 
class="cmtt-8">(0)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">{|</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">gfe</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21037r28"></a><span 
class="cmr-8">28</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Rational</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Rational</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">gfe</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">population_frequency</span><span 
class="cmtt-8">))</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21038r29"></a><span 
class="cmr-8">29</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">}</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1.0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21039r30"></a><span 
class="cmr-8">30</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">raise</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">ArgumentError</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">frequencies</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genomes</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">must</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21040r31"></a><span 
class="cmr-8">31</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21041r32"></a><span 
class="cmr-8">32</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">starting_population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">max_population</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21042r33"></a><span 
class="cmr-8">33</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">raise</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">ArgumentError</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">starting</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">must</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">be</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">less</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">than</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21043r34"></a><span 
class="cmr-8">34</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">maximum</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">population</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21044r35"></a><span 
class="cmr-8">35</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21045r36"></a><span 
class="cmr-8">36</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21046r37"></a><span 
class="cmr-8">37</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Set</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environments</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">parameters</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21047r38"></a><span 
class="cmr-8">38</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@temperature</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21048r39"></a><span 
class="cmr-8">39</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@max_population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">max_population</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21049r40"></a><span 
class="cmr-8">40</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">[]</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21050r41"></a><span 
class="cmr-8">41</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21051r42"></a><span 
class="cmr-8">42</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Populate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21052r43"></a><span 
class="cmr-8">43</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genomes_for_environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">each</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">genome_for_species</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21053r44"></a><span 
class="cmr-8">44</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">starting_population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome_for_species</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">population_frequency</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21054r45"></a><span 
class="cmr-8">45</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">round</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">times</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21055r46"></a><span 
class="cmr-8">46</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome_for_species</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">dup</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21056r47"></a><span 
class="cmr-8">47</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rand</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21057r48"></a><span 
class="cmr-8">48</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Organism</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">self</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21058r49"></a><span 
class="cmr-8">49</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21059r50"></a><span 
class="cmr-8">50</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21060r51"></a><span 
class="cmr-8">51</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21061r52"></a><span 
class="cmr-8">52</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21062r53"></a><span 
class="cmr-8">53</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Runs</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">steps</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">default</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1000).</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21063r54"></a><span 
class="cmr-8">54</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">run</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">=1000)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21064r55"></a><span 
class="cmr-8">55</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">times</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">{</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">step</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">}</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21065r56"></a><span 
class="cmr-8">56</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21066r57"></a><span 
class="cmr-8">57</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21067r58"></a><span 
class="cmr-8">58</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Step</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">each</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">then</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">calculate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">culling</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">probability</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">based</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21068r59"></a><span 
class="cmr-8">59</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">current</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">fraction</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">maximum</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">population</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">probability</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21069r60"></a><span 
class="cmr-8">60</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">calculated</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">$</span><span 
class="cmtt-8">\</span><span 
class="cmtt-8">frac</span><span 
class="cmtt-8">{1}{(</span><span 
class="cmtt-8">N</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">n</span><span 
class="cmtt-8">)+1}</span><span 
class="cmtt-8">$</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">where</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">$N$</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">maximum</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">population</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21070r61"></a><span 
class="cmr-8">61</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">$n$</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">current</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">number</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21071r62"></a><span 
class="cmr-8">62</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">This</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">probability</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">applied</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">itteratively</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">until</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">no</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21072r63"></a><span 
class="cmr-8">63</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">culled</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21073r64"></a><span 
class="cmr-8">64</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">step</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21074r65"></a><span 
class="cmr-8">65</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">each</span><span 
class="cmtt-8">(&amp;:</span><span 
class="cmtt-8">step</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21075r66"></a><span 
class="cmr-8">66</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21076r67"></a><span 
class="cmr-8">67</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">while</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">rand</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(1.0</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">((</span><span 
class="cmtt-8">@max_population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1)))</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21077r68"></a><span 
class="cmr-8">68</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">delete_at</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">rand</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">))</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21078r69"></a><span 
class="cmr-8">69</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21079r70"></a><span 
class="cmr-8">70</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21080r71"></a><span 
class="cmr-8">71</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21081r72"></a><span 
class="cmr-8">72</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add_organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">attempts</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">an</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21082r73"></a><span 
class="cmr-8">73</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">If</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">there</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">adequate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">capacity</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">added</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">returns</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21083r74"></a><span 
class="cmr-8">74</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">true</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">If</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">currently</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">full</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">then</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nothing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">done</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21084r75"></a><span 
class="cmr-8">75</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">returns</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">false</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21085r76"></a><span 
class="cmr-8">76</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add_organism</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21086r77"></a><span 
class="cmr-8">77</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@max_population</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21087r78"></a><span 
class="cmr-8">78</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21088r79"></a><span 
class="cmr-8">79</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">true</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21089r80"></a><span 
class="cmr-8">80</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">else</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21090r81"></a><span 
class="cmr-8">81</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">false</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21091r82"></a><span 
class="cmr-8">82</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21092r83"></a><span 
class="cmr-8">83</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21093r84"></a><span 
class="cmr-8">84</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21094r85"></a><span 
class="cmr-8">85</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">returns</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">hash</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">containing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">values</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">this</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21095r86"></a><span 
class="cmr-8">86</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">well</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">results</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">iterating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">over</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21096r87"></a><span 
class="cmr-8">87</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">array</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">calling</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">each</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">s</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">turn</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21097r88"></a><span 
class="cmr-8">88</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21098r89"></a><span 
class="cmr-8">89</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">{</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@temperature</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21099r90"></a><span 
class="cmr-8">90</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">max_population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@max_population</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21100r91"></a><span 
class="cmr-8">91</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">current_population</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21101r92"></a><span 
class="cmr-8">92</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">collect</span><span 
class="cmtt-8">{|</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">}</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">}</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21102r93"></a><span 
class="cmr-8">93</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21103r94"></a><span 
class="cmr-8">94</span></span><span 
class="cmtt-8">end</span>
       </div>
<!--l. 11--><div class="lstinputlisting">
<br /> <div class="caption" 
><span class="id">: </span><span  
class="content">organism.rb</span></div><!--tex4ht:label?: x1-21000A --><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21104r1"></a><span 
class="cmr-8">1</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Copyright</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">c</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">2009</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Joshua</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Ballanco</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21105r2"></a><span 
class="cmr-8">2</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21106r3"></a><span 
class="cmr-8">3</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Organism</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21107r4"></a><span 
class="cmr-8">4</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21108r5"></a><span 
class="cmr-8">5</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Abstract</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">This</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">base</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">evolving</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">It</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">represents</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">an</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21109r6"></a><span 
class="cmr-8">6</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">abstract</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">which</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">its</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">using</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">DNA</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21110r7"></a><span 
class="cmr-8">7</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">when</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">finished</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">making</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">copy</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">into</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">two</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organims</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21111r8"></a><span 
class="cmr-8">8</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21112r9"></a><span 
class="cmr-8">9</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">implemented</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">finite</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">state</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">machine</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">with</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21113r10"></a><span 
class="cmr-8">10</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">following</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">states</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21114r11"></a><span 
class="cmr-8">11</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">--</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicate_genome</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">synthesizing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">using</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">its</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21115r12"></a><span 
class="cmr-8">12</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">existing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">template</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21116r13"></a><span 
class="cmr-8">13</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">--</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">divide</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Split</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">into</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">two</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">adding</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21117r14"></a><span 
class="cmr-8">14</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">capacity</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">it</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">exists</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21118r15"></a><span 
class="cmr-8">15</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21119r16"></a><span 
class="cmr-8">16</span></span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Organism</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21120r17"></a><span 
class="cmr-8">17</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Initialization</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">consists</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">setting</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">translating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21121r18"></a><span 
class="cmr-8">18</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">from</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">passing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reference</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">which</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21122r19"></a><span 
class="cmr-8">19</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">this</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">lives</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21123r20"></a><span 
class="cmr-8">20</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initialize</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21124r21"></a><span 
class="cmr-8">21</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21125r22"></a><span 
class="cmr-8">22</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21126r23"></a><span 
class="cmr-8">23</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">translate_polymerase</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21127r24"></a><span 
class="cmr-8">24</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21128r25"></a><span 
class="cmr-8">25</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">We</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">ll</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">start</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">with</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21129r26"></a><span 
class="cmr-8">26</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@next_step</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">(:</span><span 
class="cmtt-8">replicate_genome</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">to_proc</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21130r27"></a><span 
class="cmr-8">27</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21131r28"></a><span 
class="cmr-8">28</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21132r29"></a><span 
class="cmr-8">29</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Step</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">this</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">set</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">next</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">step</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">value</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21133r30"></a><span 
class="cmr-8">30</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">step</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21134r31"></a><span 
class="cmr-8">31</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@next_step</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@next_step</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">call</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21135r32"></a><span 
class="cmr-8">32</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">self</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21136r33"></a><span 
class="cmr-8">33</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21137r34"></a><span 
class="cmr-8">34</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21138r35"></a><span 
class="cmr-8">35</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">We</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">re</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">middle</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">creating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">To</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">this</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">allow</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21139r36"></a><span 
class="cmr-8">36</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">many</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">it</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">will</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Following</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">query</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21140r37"></a><span 
class="cmr-8">37</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">its</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">current</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">status</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">proceed</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">based</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21141r38"></a><span 
class="cmr-8">38</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">information</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21142r39"></a><span 
class="cmr-8">39</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicate_genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21143r40"></a><span 
class="cmr-8">40</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">add_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21144r41"></a><span 
class="cmr-8">41</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">case</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">status</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21145r42"></a><span 
class="cmr-8">42</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">when</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">polymerizing</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21146r43"></a><span 
class="cmr-8">43</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">(:</span><span 
class="cmtt-8">replicate_genome</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">to_proc</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21147r44"></a><span 
class="cmr-8">44</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">when</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">finished</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21148r45"></a><span 
class="cmr-8">45</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">(:</span><span 
class="cmtt-8">divide</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">to_proc</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21149r46"></a><span 
class="cmr-8">46</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21150r47"></a><span 
class="cmr-8">47</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21151r48"></a><span 
class="cmr-8">48</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21152r49"></a><span 
class="cmr-8">49</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">In</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">order</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">divide</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">first</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">extract</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">If</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">has</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21153r50"></a><span 
class="cmr-8">50</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">too</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">many</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">errors</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">then</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">try</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">again</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Otherwise</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">create</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21154r51"></a><span 
class="cmr-8">51</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">from</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">attempt</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">insert</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">into</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21155r52"></a><span 
class="cmr-8">52</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">If</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">there</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">no</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">room</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">keep</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">trying</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">until</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">there</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">or</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">are</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21156r53"></a><span 
class="cmr-8">53</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">randomly</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">killed</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Once</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">ve</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">put</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21157r54"></a><span 
class="cmr-8">54</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">start</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">again</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21158r55"></a><span 
class="cmr-8">55</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">divide</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21159r56"></a><span 
class="cmr-8">56</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@new_genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">||=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new_finished_genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21160r57"></a><span 
class="cmr-8">57</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">unless</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@new_genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21161r58"></a><span 
class="cmr-8">58</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21162r59"></a><span 
class="cmr-8">59</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">(:</span><span 
class="cmtt-8">replicate_genome</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">to_proc</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21163r60"></a><span 
class="cmr-8">60</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21164r61"></a><span 
class="cmr-8">61</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21165r62"></a><span 
class="cmr-8">62</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@new_organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">||=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Organism</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@new_genome</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@environment</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21166r63"></a><span 
class="cmr-8">63</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">add_organism</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@new_organism</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21167r64"></a><span 
class="cmr-8">64</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@new_genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nil</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21168r65"></a><span 
class="cmr-8">65</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@new_organism</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nil</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21169r66"></a><span 
class="cmr-8">66</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21170r67"></a><span 
class="cmr-8">67</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">(:</span><span 
class="cmtt-8">replicate_genome</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">to_proc</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21171r68"></a><span 
class="cmr-8">68</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21172r69"></a><span 
class="cmr-8">69</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21173r70"></a><span 
class="cmr-8">70</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">(:</span><span 
class="cmtt-8">divide</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">to_proc</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21174r71"></a><span 
class="cmr-8">71</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21175r72"></a><span 
class="cmr-8">72</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21176r73"></a><span 
class="cmr-8">73</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">returns</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">details</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">about</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">s</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21177r74"></a><span 
class="cmr-8">74</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21178r75"></a><span 
class="cmr-8">75</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21179r76"></a><span 
class="cmr-8">76</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">{</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21180r77"></a><span 
class="cmr-8">77</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">}</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21181r78"></a><span 
class="cmr-8">78</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21182r79"></a><span 
class="cmr-8">79</span></span><span 
class="cmtt-8">end</span>
       </div>
<!--l. 13--><div class="lstinputlisting">
<br /> <div class="caption" 
><span class="id">: </span><span  
class="content">genome.rb</span></div><!--tex4ht:label?: x1-21000A --><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21183r1"></a><span 
class="cmr-8">1</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Copyright</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">c</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">2009-2011</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Joshua</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Ballanco</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21184r2"></a><span 
class="cmr-8">2</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21185r3"></a><span 
class="cmr-8">3</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21186r4"></a><span 
class="cmr-8">4</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21187r5"></a><span 
class="cmr-8">5</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Abstract</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">comprises</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">an</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">It</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21188r6"></a><span 
class="cmr-8">6</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initialized</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">with</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">some</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">information</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">about</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">sort</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21189r7"></a><span 
class="cmr-8">7</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">enzyme</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">it</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">codes</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">As</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">it</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">copied</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">by</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">),</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">it</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">tracks</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21190r8"></a><span 
class="cmr-8">8</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">progress</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">count</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">made</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">When</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">requested</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">can</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21191r9"></a><span 
class="cmr-8">9</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">generate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">or</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">copy</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">itself</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21192r10"></a><span 
class="cmr-8">10</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21193r11"></a><span 
class="cmr-8">11</span></span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21194r12"></a><span 
class="cmr-8">12</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">attr_reader</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">errors</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21195r13"></a><span 
class="cmr-8">13</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">attr_accessor</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21196r14"></a><span 
class="cmr-8">14</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21197r15"></a><span 
class="cmr-8">15</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Initialization</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">requires</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">well</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21198r16"></a><span 
class="cmr-8">16</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">coded</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">by</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21199r17"></a><span 
class="cmr-8">17</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initialize</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase_rate</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">mutable</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">true</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21200r18"></a><span 
class="cmr-8">18</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@length</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21201r19"></a><span 
class="cmr-8">19</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21202r20"></a><span 
class="cmr-8">20</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21203r21"></a><span 
class="cmr-8">21</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@mutable</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">mutable</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21204r22"></a><span 
class="cmr-8">22</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21205r23"></a><span 
class="cmr-8">23</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21206r24"></a><span 
class="cmr-8">24</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21207r25"></a><span 
class="cmr-8">25</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21208r26"></a><span 
class="cmr-8">26</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">This</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">gets</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">called</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">during</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Genome</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">dup</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">original</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21209r27"></a><span 
class="cmr-8">27</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">left</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">unchanged</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">but</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">properties</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">generated</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">are</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21210r28"></a><span 
class="cmr-8">28</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">recalculated</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">based</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">how</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">many</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">mutations</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">occurred</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">during</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replication</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21211r29"></a><span 
class="cmr-8">29</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initialize_copy</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">orig</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21212r30"></a><span 
class="cmr-8">30</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@mutable</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21213r31"></a><span 
class="cmr-8">31</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">high_dev</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">MAX_POLY_RATE</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21214r32"></a><span 
class="cmr-8">32</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">low_dev</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">MIN_POLY_RATE</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21215r33"></a><span 
class="cmr-8">33</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">max_dev</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">MAX_POLY_RATE</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">MIN_POLY_RATE</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">2.0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21216r34"></a><span 
class="cmr-8">34</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">mut_frac</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@length</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">to_f</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">MAX_TOL_MUT_RATE</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21217r35"></a><span 
class="cmr-8">35</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">mut_frac</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">mut_frac</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">?</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">mut_frac</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21218r36"></a><span 
class="cmr-8">36</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@change_in_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">mut_frac</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">max_dev</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">round</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21219r37"></a><span 
class="cmr-8">37</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21220r38"></a><span 
class="cmr-8">38</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@change_in_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">high_dev</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21221r39"></a><span 
class="cmr-8">39</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@change_in_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21222r40"></a><span 
class="cmr-8">40</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">elsif</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@change_in_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">low_dev</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21223r41"></a><span 
class="cmr-8">41</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@change_in_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21224r42"></a><span 
class="cmr-8">42</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">elsif</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rand</span><span 
class="cmtt-8">(2)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21225r43"></a><span 
class="cmr-8">43</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@change_in_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21226r44"></a><span 
class="cmr-8">44</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">else</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21227r45"></a><span 
class="cmr-8">45</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@change_in_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21228r46"></a><span 
class="cmr-8">46</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21229r47"></a><span 
class="cmr-8">47</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21230r48"></a><span 
class="cmr-8">48</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21231r49"></a><span 
class="cmr-8">49</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">MAX_POLY_RATE</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">||</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">MIN_POLY_RATE</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21232r50"></a><span 
class="cmr-8">50</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">raise</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">RuntimeError</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">Polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">out</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Bounds</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21233r51"></a><span 
class="cmr-8">51</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21234r52"></a><span 
class="cmr-8">52</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21235r53"></a><span 
class="cmr-8">53</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">At</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21236r54"></a><span 
class="cmr-8">54</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">self</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21237r55"></a><span 
class="cmr-8">55</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21238r56"></a><span 
class="cmr-8">56</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21239r57"></a><span 
class="cmr-8">57</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Start</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">over</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicating</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">e</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">g</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">an</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">unviable</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">copy</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">was</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">made</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21240r58"></a><span 
class="cmr-8">58</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">discarded</span><span 
class="cmtt-8">).</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21241r59"></a><span 
class="cmr-8">59</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21242r60"></a><span 
class="cmr-8">60</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21243r61"></a><span 
class="cmr-8">61</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21244r62"></a><span 
class="cmr-8">62</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21245r63"></a><span 
class="cmr-8">63</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21246r64"></a><span 
class="cmr-8">64</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Add</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotide</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replica</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">If</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">there</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">was</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">an</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">erroneous</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21247r65"></a><span 
class="cmr-8">65</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">inclusion</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">number</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">well</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21248r66"></a><span 
class="cmr-8">66</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add_nucleotide</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">error</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">false</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21249r67"></a><span 
class="cmr-8">67</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">error</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21250r68"></a><span 
class="cmr-8">68</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21251r69"></a><span 
class="cmr-8">69</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21252r70"></a><span 
class="cmr-8">70</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21253r71"></a><span 
class="cmr-8">71</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Seed</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">with</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">defined</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">by</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">this</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21254r72"></a><span 
class="cmr-8">72</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21255r73"></a><span 
class="cmr-8">73</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">translate_polymerase</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21256r74"></a><span 
class="cmr-8">74</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">self</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@directionality</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@polymerase_rate</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21257r75"></a><span 
class="cmr-8">75</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21258r76"></a><span 
class="cmr-8">76</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21259r77"></a><span 
class="cmr-8">77</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">This</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">returns</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">viability</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replica</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">being</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">created</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21260r78"></a><span 
class="cmr-8">78</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">considered</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">viable</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">so</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">long</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">as</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">number</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">added</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21261r79"></a><span 
class="cmr-8">79</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">meets</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">or</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">exceeds</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21262r80"></a><span 
class="cmr-8">80</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">viable</span><span 
class="cmtt-8">?</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21263r81"></a><span 
class="cmr-8">81</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@length</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21264r82"></a><span 
class="cmr-8">82</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">true</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21265r83"></a><span 
class="cmr-8">83</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">else</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21266r84"></a><span 
class="cmr-8">84</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">false</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21267r85"></a><span 
class="cmr-8">85</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21268r86"></a><span 
class="cmr-8">86</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21269r87"></a><span 
class="cmr-8">87</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21270r88"></a><span 
class="cmr-8">88</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">returns</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">hash</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">properties</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">about</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21271r89"></a><span 
class="cmr-8">89</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21272r90"></a><span 
class="cmr-8">90</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">{</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@length</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21273r91"></a><span 
class="cmr-8">91</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">change_in_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@change_in_rate</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21274r92"></a><span 
class="cmr-8">92</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@added_nucleotides</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21275r93"></a><span 
class="cmr-8">93</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@errors</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">}</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21276r94"></a><span 
class="cmr-8">94</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21277r95"></a><span 
class="cmr-8">95</span></span><span 
class="cmtt-8">end</span>
       </div>
<!--l. 15--><div class="lstinputlisting">
<br /> <div class="caption" 
><span class="id">: </span><span  
class="content">polymerase.rb</span></div><!--tex4ht:label?: x1-21000A --><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21278r1"></a><span 
class="cmr-8">1</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Copyright</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">c</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">2009-2010</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Joshua</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Ballanco</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21279r2"></a><span 
class="cmr-8">2</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21280r3"></a><span 
class="cmr-8">3</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Polymerase</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21281r4"></a><span 
class="cmr-8">4</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21282r5"></a><span 
class="cmr-8">5</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Abstract</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">describes</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">generic</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotide</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21283r6"></a><span 
class="cmr-8">6</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">When</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">asked</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">it</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">will</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">number</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nascient</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21284r7"></a><span 
class="cmr-8">7</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">insertion</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">depends</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21285r8"></a><span 
class="cmr-8">8</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">was</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initialized</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">with</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Each</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">will</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">always</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">be</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">in</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">one</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21286r9"></a><span 
class="cmr-8">9</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">two</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">states</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21287r10"></a><span 
class="cmr-8">10</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">--</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerizing</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">not</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">yet</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">finished</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21288r11"></a><span 
class="cmr-8">11</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">--</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">finished</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">has</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">been</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">completely</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicated</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21289r12"></a><span 
class="cmr-8">12</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21290r13"></a><span 
class="cmr-8">13</span></span><span 
class="cmtt-8">class</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Polymerase</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21291r14"></a><span 
class="cmr-8">14</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">attr_reader</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">status</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21292r15"></a><span 
class="cmr-8">15</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21293r16"></a><span 
class="cmr-8">16</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">created</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">must</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">specify</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21294r17"></a><span 
class="cmr-8">17</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerization</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simulation</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">when</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">it</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">created</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21295r18"></a><span 
class="cmr-8">18</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">initialize</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21296r19"></a><span 
class="cmr-8">19</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@status</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">polymerizing</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21297r20"></a><span 
class="cmr-8">20</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21298r21"></a><span 
class="cmr-8">21</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21299r22"></a><span 
class="cmr-8">22</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">unless</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">forward</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">||</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">reverse</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21300r23"></a><span 
class="cmr-8">23</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">raise</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">ArgumentError</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">must</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">be</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">forward</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">or</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">reverse</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21301r24"></a><span 
class="cmr-8">24</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21302r25"></a><span 
class="cmr-8">25</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21303r26"></a><span 
class="cmr-8">26</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@temperature</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21304r27"></a><span 
class="cmr-8">27</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21305r28"></a><span 
class="cmr-8">28</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21306r29"></a><span 
class="cmr-8">29</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">This</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">adds</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotide</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">First</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">checks</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21307r30"></a><span 
class="cmr-8">30</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">see</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">already</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">finished</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">being</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replicated</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">If</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">not</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">then</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">it</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21308r31"></a><span 
class="cmr-8">31</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">will</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">number</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">up</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">this</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">When</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21309r32"></a><span 
class="cmr-8">32</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">adding</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotides</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">there</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">chance</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nucleotide</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">being</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">added</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21310r33"></a><span 
class="cmr-8">33</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">has</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">become</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">dephosphorylated</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">There</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">also</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">chance</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">dependent</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21311r34"></a><span 
class="cmr-8">34</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">an</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">error</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">made</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">during</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">replication</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21312r35"></a><span 
class="cmr-8">35</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">add_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21313r36"></a><span 
class="cmr-8">36</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">added_nucleotides</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21314r37"></a><span 
class="cmr-8">37</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@status</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">finished</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21315r38"></a><span 
class="cmr-8">38</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21316r39"></a><span 
class="cmr-8">39</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21317r40"></a><span 
class="cmr-8">40</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21318r41"></a><span 
class="cmr-8">41</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@rate</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">times</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21319r42"></a><span 
class="cmr-8">42</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">thermal_prob</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Math</span><span 
class="cmtt-8">::</span><span 
class="cmtt-8">E</span><span 
class="cmtt-8">**(-1.0</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@temperature</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21320r43"></a><span 
class="cmr-8">43</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">((</span><span 
class="cmtt-8">@directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">forward</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&amp;&amp;</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21321r44"></a><span 
class="cmr-8">44</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">rand</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">thermal_prob</span><span 
class="cmtt-8">**2))</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21322r45"></a><span 
class="cmr-8">45</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">next</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21323r46"></a><span 
class="cmr-8">46</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">elsif</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">((</span><span 
class="cmtt-8">@directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">reverse</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&amp;&amp;</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21324r47"></a><span 
class="cmr-8">47</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">rand</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">thermal_prob</span><span 
class="cmtt-8">**2</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">MAX_POLY_RATE</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1))))</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21325r48"></a><span 
class="cmr-8">48</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21326r49"></a><span 
class="cmr-8">49</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">else</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21327r50"></a><span 
class="cmr-8">50</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">add_nucleotide</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">rand</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">thermal_prob</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Math</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">sqrt</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@rate</span><span 
class="cmtt-8">)))</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21328r51"></a><span 
class="cmr-8">51</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21329r52"></a><span 
class="cmr-8">52</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21330r53"></a><span 
class="cmr-8">53</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21331r54"></a><span 
class="cmr-8">54</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21332r55"></a><span 
class="cmr-8">55</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">If</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">re</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">done</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerizing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">that</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">ve</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">produced</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">viable</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21333r56"></a><span 
class="cmr-8">56</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">then</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Otherwise</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">return</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nil</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21334r57"></a><span 
class="cmr-8">57</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">new_finished_genome</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21335r58"></a><span 
class="cmr-8">58</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@status</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">finished</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&amp;&amp;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">viable</span><span 
class="cmtt-8">?</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21336r59"></a><span 
class="cmr-8">59</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">dup</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21337r60"></a><span 
class="cmr-8">60</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">else</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21338r61"></a><span 
class="cmr-8">61</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">nil</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21339r62"></a><span 
class="cmr-8">62</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21340r63"></a><span 
class="cmr-8">63</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21341r64"></a><span 
class="cmr-8">64</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21342r65"></a><span 
class="cmr-8">65</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Resets</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">back</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">starting</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">with</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">fresh</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">start</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21343r66"></a><span 
class="cmr-8">66</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerizing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">anew</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21344r67"></a><span 
class="cmr-8">67</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21345r68"></a><span 
class="cmr-8">68</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">reset</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21346r69"></a><span 
class="cmr-8">69</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@status</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">polymerizing</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21347r70"></a><span 
class="cmr-8">70</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21348r71"></a><span 
class="cmr-8">71</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21349r72"></a><span 
class="cmr-8">72</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">The</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">method</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">returns</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">hash</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">properties</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">about</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21350r73"></a><span 
class="cmr-8">73</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21351r74"></a><span 
class="cmr-8">74</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">{</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">status</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@status</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21352r75"></a><span 
class="cmr-8">75</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@directionality</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21353r76"></a><span 
class="cmr-8">76</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">}</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21354r77"></a><span 
class="cmr-8">77</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21355r78"></a><span 
class="cmr-8">78</span></span><span 
class="cmtt-8">end</span>
       </div>
<!--l. 17--><div class="lstinputlisting">
<br /> <div class="caption" 
><span class="id">: </span><span  
class="content">evolver</span></div><!--tex4ht:label?: x1-21000A --><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21356r1"></a><span 
class="cmr-8">1</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">!/</span><span 
class="cmtt-8">usr</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">bin</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">ruby</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21357r2"></a><span 
class="cmr-8">2</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21358r3"></a><span 
class="cmr-8">3</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Copyright</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">c</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">2009-2011</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Joshua</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Ballanco</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21359r4"></a><span 
class="cmr-8">4</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21360r5"></a><span 
class="cmr-8">5</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">evolver</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">This</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">is</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">command</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">line</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">utility</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">manipulate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Evolver</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">model</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21361r6"></a><span 
class="cmr-8">6</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21362r7"></a><span 
class="cmr-8">7</span></span><span 
class="cmtt-8">$LOAD_PATH</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">File</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">join</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">File</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">dirname</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">__FILE__</span><span 
class="cmtt-8">),</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">..</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">lib</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21363r8"></a><span 
class="cmr-8">8</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21364r9"></a><span 
class="cmr-8">9</span></span><span 
class="cmtt-8">require</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">optparse</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21365r10"></a><span 
class="cmr-8">10</span></span><span 
class="cmtt-8">require</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">yaml</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21366r11"></a><span 
class="cmr-8">11</span></span><span 
class="cmtt-8">require</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">evolver</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">constants</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21367r12"></a><span 
class="cmr-8">12</span></span><span 
class="cmtt-8">require</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">evolver</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">organism</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21368r13"></a><span 
class="cmr-8">13</span></span><span 
class="cmtt-8">require</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">evolver</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21369r14"></a><span 
class="cmr-8">14</span></span><span 
class="cmtt-8">require</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">evolver</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21370r15"></a><span 
class="cmr-8">15</span></span><span 
class="cmtt-8">require</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">evolver</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21371r16"></a><span 
class="cmr-8">16</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21372r17"></a><span 
class="cmr-8">17</span></span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">{}</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21373r18"></a><span 
class="cmr-8">18</span></span><span 
class="cmtt-8">OptionParser</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">opts</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21374r19"></a><span 
class="cmr-8">19</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">opts</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">banner</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">Usage</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">evolver</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">[</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">file</span><span 
class="cmtt-8">&#x003E;.</span><span 
class="cmtt-8">yml</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21375r20"></a><span 
class="cmr-8">20</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21376r21"></a><span 
class="cmr-8">21</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1000</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21377r22"></a><span 
class="cmr-8">22</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">opts</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">n</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">[</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21378r23"></a><span 
class="cmr-8">23</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">Number</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">run</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">per</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">iter</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21379r24"></a><span 
class="cmr-8">24</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">iter</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">to_i</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21380r25"></a><span 
class="cmr-8">25</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21381r26"></a><span 
class="cmr-8">26</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21382r27"></a><span 
class="cmr-8">27</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">10</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21383r28"></a><span 
class="cmr-8">28</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">opts</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">f</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">[</span><span 
class="cmtt-8">frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21384r29"></a><span 
class="cmr-8">29</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">Report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">frequency</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">between</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reports</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">freq</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21385r30"></a><span 
class="cmr-8">30</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">freq</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">to_i</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21386r31"></a><span 
class="cmr-8">31</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21387r32"></a><span 
class="cmr-8">32</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21388r33"></a><span 
class="cmr-8">33</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">snapshot_frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21389r34"></a><span 
class="cmr-8">34</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">opts</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">s</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">[</span><span 
class="cmtt-8">frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21390r35"></a><span 
class="cmr-8">35</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">Snapshot</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">frequency</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">between</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">snapshots</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">freq</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21391r36"></a><span 
class="cmr-8">36</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">snapshot_frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">freq</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">to_i</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21392r37"></a><span 
class="cmr-8">37</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21393r38"></a><span 
class="cmr-8">38</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21394r39"></a><span 
class="cmr-8">39</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">threads</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21395r40"></a><span 
class="cmr-8">40</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">opts</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">j</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">[</span><span 
class="cmtt-8">threads</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21396r41"></a><span 
class="cmr-8">41</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">Number</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">of</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">concurrent</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">threads</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">run</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">calculations</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">thr</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21397r42"></a><span 
class="cmr-8">42</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">threads</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">thr</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">to_i</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21398r43"></a><span 
class="cmr-8">43</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21399r44"></a><span 
class="cmr-8">44</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21400r45"></a><span 
class="cmr-8">45</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_values</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">%</span><span 
class="cmtt-8">W</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">forward_num</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21401r46"></a><span 
class="cmr-8">46</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reverse_num</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21402r47"></a><span 
class="cmr-8">47</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">forward_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21403r48"></a><span 
class="cmr-8">48</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reverse_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21404r49"></a><span 
class="cmr-8">49</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism_num</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21405r50"></a><span 
class="cmr-8">50</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">organism_rate</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21406r51"></a><span 
class="cmr-8">51</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">mutation_rate</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21407r52"></a><span 
class="cmr-8">52</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">opts</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">on</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">r</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">[</span><span 
class="cmtt-8">Value1</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">Value2</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">Value3</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Array</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21408r53"></a><span 
class="cmr-8">53</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">Values</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">to</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">values</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21409r54"></a><span 
class="cmr-8">54</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_values</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">values</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21410r55"></a><span 
class="cmr-8">55</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21411r56"></a><span 
class="cmr-8">56</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">parse</span><span 
class="cmtt-8">!</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21412r57"></a><span 
class="cmr-8">57</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21413r58"></a><span 
class="cmr-8">58</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">run</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21414r59"></a><span 
class="cmr-8">59</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genomes_for_env</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">genomes</span><span 
class="cmtt-8">].</span><span 
class="cmtt-8">collect</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21415r60"></a><span 
class="cmr-8">60</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">GenomeForSpecies</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">Genome</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">length</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21416r61"></a><span 
class="cmr-8">61</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">polymerase_rate</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21417r62"></a><span 
class="cmr-8">62</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21418r63"></a><span 
class="cmr-8">63</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">mutable_polymerase</span><span 
class="cmtt-8">]),</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21419r64"></a><span 
class="cmr-8">64</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">freq</span><span 
class="cmtt-8">])</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21420r65"></a><span 
class="cmr-8">65</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21421r66"></a><span 
class="cmr-8">66</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Environment</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">new</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">temperature</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21422r67"></a><span 
class="cmr-8">67</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">max_population</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21423r68"></a><span 
class="cmr-8">68</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">starting_population</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21424r69"></a><span 
class="cmr-8">69</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">genomes_for_env</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21425r70"></a><span 
class="cmr-8">70</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21426r71"></a><span 
class="cmr-8">71</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">use_threads</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">threads</span><span 
class="cmtt-8">])</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">threads</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21427r72"></a><span 
class="cmr-8">72</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Run</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simulation</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">frequency</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">then</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">output</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">requested</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21428r73"></a><span 
class="cmr-8">73</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">values</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Keep</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">doing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">this</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">until</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">we</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">ve</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">hit</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">max</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">limit</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21429r74"></a><span 
class="cmr-8">74</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@iterations_complete</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21430r75"></a><span 
class="cmr-8">75</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">name</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">true</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_values</span><span 
class="cmtt-8">])</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21431r76"></a><span 
class="cmr-8">76</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">while</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">@iterations_complete</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">])</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21432r77"></a><span 
class="cmr-8">77</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">run</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_frequency</span><span 
class="cmtt-8">])</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21433r78"></a><span 
class="cmr-8">78</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">name</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">false</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_values</span><span 
class="cmtt-8">])</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21434r79"></a><span 
class="cmr-8">79</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">snapshot_frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003E;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&amp;&amp;</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21435r80"></a><span 
class="cmr-8">80</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@iterations_complete</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">%</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">snapshot_frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21436r81"></a><span 
class="cmr-8">81</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">File</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">open</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">name</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">_snapshots</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">txt</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">snapshotfile</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21437r82"></a><span 
class="cmr-8">82</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">snapshotfile</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">YAML</span><span 
class="cmtt-8">::</span><span 
class="cmtt-8">dump</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">...\</span><span 
class="cmtt-8">n</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21438r83"></a><span 
class="cmr-8">83</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21439r84"></a><span 
class="cmr-8">84</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21440r85"></a><span 
class="cmr-8">85</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@iterations_complete</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_frequency</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21441r86"></a><span 
class="cmr-8">86</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21442r87"></a><span 
class="cmr-8">87</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">run</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">iterations</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">-</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">@iterations_complete</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21443r88"></a><span 
class="cmr-8">88</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">name</span><span 
class="cmtt-8">],</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">false</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">report_values</span><span 
class="cmtt-8">])</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21444r89"></a><span 
class="cmr-8">89</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21445r90"></a><span 
class="cmr-8">90</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21446r91"></a><span 
class="cmr-8">91</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">name</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">print_header</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">*</span><span 
class="cmtt-8">report_values</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21447r92"></a><span 
class="cmr-8">92</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">File</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">open</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">name</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">_out</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">csv</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">a</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">outfile</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21448r93"></a><span 
class="cmr-8">93</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">outfile</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_values</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">join</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">\</span><span 
class="cmtt-8">n</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">if</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">print_header</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21449r94"></a><span 
class="cmr-8">94</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">report</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21450r95"></a><span 
class="cmr-8">95</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">values</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_values</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">collect</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">value</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21451r96"></a><span 
class="cmr-8">96</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">send</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">report_</span><span 
class="cmtt-8">#{</span><span 
class="cmtt-8">value</span><span 
class="cmtt-8">}</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">to_sym</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21452r97"></a><span 
class="cmr-8">97</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21453r98"></a><span 
class="cmr-8">98</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">outfile</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">values</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">join</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#8217;</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x003C;&#x003C;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">\</span><span 
class="cmtt-8">n</span><span 
class="cmtt-8">"</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21454r99"></a><span 
class="cmr-8">99</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21455r100"></a><span 
class="cmr-8">100</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21456r101"></a><span 
class="cmr-8">101</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21457r102"></a><span 
class="cmr-8">102</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Methods</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">reporting</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simulation</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">parameters</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21458r103"></a><span 
class="cmr-8">103</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_forward_num</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21459r104"></a><span 
class="cmr-8">104</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">].</span><span 
class="cmtt-8">inject</span><span 
class="cmtt-8">(0)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21460r105"></a><span 
class="cmr-8">105</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">][:</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">forward</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">?</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21461r106"></a><span 
class="cmr-8">106</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21462r107"></a><span 
class="cmr-8">107</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21463r108"></a><span 
class="cmr-8">108</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21464r109"></a><span 
class="cmr-8">109</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_reverse_num</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21465r110"></a><span 
class="cmr-8">110</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">].</span><span 
class="cmtt-8">inject</span><span 
class="cmtt-8">(0)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21466r111"></a><span 
class="cmr-8">111</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">][:</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">reverse</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">?</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">1</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21467r112"></a><span 
class="cmr-8">112</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21468r113"></a><span 
class="cmr-8">113</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21469r114"></a><span 
class="cmr-8">114</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21470r115"></a><span 
class="cmr-8">115</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_forward_rate</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21471r116"></a><span 
class="cmr-8">116</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">].</span><span 
class="cmtt-8">inject</span><span 
class="cmtt-8">(0.0)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21472r117"></a><span 
class="cmr-8">117</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">][:</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">forward</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">?</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21473r118"></a><span 
class="cmr-8">118</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">][:</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21474r119"></a><span 
class="cmr-8">119</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_forward_num</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21475r120"></a><span 
class="cmr-8">120</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21476r121"></a><span 
class="cmr-8">121</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21477r122"></a><span 
class="cmr-8">122</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_reverse_rate</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21478r123"></a><span 
class="cmr-8">123</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">].</span><span 
class="cmtt-8">inject</span><span 
class="cmtt-8">(0.0)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21479r124"></a><span 
class="cmr-8">124</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">][:</span><span 
class="cmtt-8">directionality</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">==</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">reverse</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">?</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21480r125"></a><span 
class="cmr-8">125</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">][:</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">0)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21481r126"></a><span 
class="cmr-8">126</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_reverse_num</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21482r127"></a><span 
class="cmr-8">127</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21483r128"></a><span 
class="cmr-8">128</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21484r129"></a><span 
class="cmr-8">129</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_organism_num</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21485r130"></a><span 
class="cmr-8">130</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">current_population</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21486r131"></a><span 
class="cmr-8">131</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21487r132"></a><span 
class="cmr-8">132</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21488r133"></a><span 
class="cmr-8">133</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_organism_rate</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21489r134"></a><span 
class="cmr-8">134</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">].</span><span 
class="cmtt-8">inject</span><span 
class="cmtt-8">(0.0)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21490r135"></a><span 
class="cmr-8">135</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">polymerase</span><span 
class="cmtt-8">][:</span><span 
class="cmtt-8">rate</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21491r136"></a><span 
class="cmr-8">136</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">current_population</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21492r137"></a><span 
class="cmr-8">137</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21493r138"></a><span 
class="cmr-8">138</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21494r139"></a><span 
class="cmr-8">139</span></span><span 
class="cmtt-8">def</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">report_mutation_rate</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21495r140"></a><span 
class="cmr-8">140</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">organisms</span><span 
class="cmtt-8">].</span><span 
class="cmtt-8">inject</span><span 
class="cmtt-8">(0.0)</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21496r141"></a><span 
class="cmr-8">141</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">total</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">+</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">org</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">genome</span><span 
class="cmtt-8">][:</span><span 
class="cmtt-8">change_in_rate</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21497r142"></a><span 
class="cmr-8">142</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">/</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">env_report</span><span 
class="cmtt-8">[:</span><span 
class="cmtt-8">current_population</span><span 
class="cmtt-8">]</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21498r143"></a><span 
class="cmr-8">143</span></span><span 
class="cmtt-8">end</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21499r144"></a><span 
class="cmr-8">144</span></span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21500r145"></a><span 
class="cmr-8">145</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">After</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">parsing</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">only</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">remaining</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">argument</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">should</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">be</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21501r146"></a><span 
class="cmr-8">146</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">specification</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">file</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">Load</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">this</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">and</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">run</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">the</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">simulation</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">for</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">each</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21502r147"></a><span 
class="cmr-8">147</span></span><span 
class="cmtt-8">#</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">described</span><span 
class="cmtt-8">:</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21503r148"></a><span 
class="cmr-8">148</span></span><span 
class="cmtt-8">environments</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">=</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">YAML</span><span 
class="cmtt-8">::</span><span 
class="cmtt-8">load_file</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">ARGV</span><span 
class="cmtt-8">[0])</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21504r149"></a><span 
class="cmr-8">149</span></span><span 
class="cmtt-8">environments</span><span 
class="cmtt-8">.</span><span 
class="cmtt-8">each</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">do</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">|</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21505r150"></a><span 
class="cmr-8">150</span></span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">run</span><span 
class="cmtt-8">(</span><span 
class="cmtt-8">options</span><span 
class="cmtt-8">,</span><span 
class="cmtt-8">&#x00A0;</span><span 
class="cmtt-8">environment</span><span 
class="cmtt-8">)</span><span 
class="cmtt-8">&#x00A0;</span><br /><span class="label"><a 
 id="x1-21506r151"></a><span 
class="cmr-8">151</span></span><span 
class="cmtt-8">end</span>
       </div>


       <h2 class="likechapterHead"><a 
 id="x1-22000A"></a>Bibliography</h2>
     <div class="thebibliography">
     <p class="bibitem" ><span class="biblabel">
  [1]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XBagshaw:1973p728"></a>C&#x00A0;R  Bagshaw  and  D&#x00A0;R  Trentham.     The  reversibility  of  adenosine
     triphosphate cleavage by myosin. <span 
class="cmti-12">The Biochemical journal</span>, 133(2):323&#8211;328, jun
     1973. Reference for the G of NTP hydrolysis.
     </p>
     <p class="bibitem" ><span class="biblabel">
  [2]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XCady:2009p718"></a>Field Cady and Hong Qian.  Open-system thermodynamic analysis of DNA
     polymerase fidelity.  <span 
class="cmti-12">Physical biology</span>, 6(3):36011, jan 2009.  Reference for error
     rate as a function of shape.
     </p>
     <p class="bibitem" ><span class="biblabel">
  [3]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XCastro:2009p953"></a>Christian   Castro,   Eric&#x00A0;D   Smidansky,   Jamie&#x00A0;J   Arnold,   Kenneth&#x00A0;R
     Maksimchuk,  Ibrahim  Moustafa,  Akira  Uchida,  Matthias  Götte,  William
     Konigsberg, and Craig&#x00A0;E Cameron. Nucleic acid polymerases use a general acid
     for nucleotidyl transfer.  <span 
class="cmti-12">Nature structural &amp; molecular biology</span>, 16(2):212&#8211;218,
     feb 2009. Reference for polymerase active site.
     </p>

     <p class="bibitem" ><span class="biblabel">
  [4]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XDarwin:1883p935"></a>C&#x00A0;Darwin.  On the origin of species: by means of natural selection, or, the
     preservation &#x2026;. <span 
class="cmti-12">books.google.com</span>, jan 1883.
     </p>
     <p class="bibitem" ><span class="biblabel">
  [5]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XFerenci:1999p703"></a>T&#x00A0;Ferenci. &#8217;Growth of bacterial cultures&#8217; 50 years on: towards an uncertainty
     principle  instead  of  constants  in  bacterial  growth  kinetics.     <span 
class="cmti-12">Research  in</span>
     <span 
class="cmti-12">microbiology</span>, 150(7):431&#8211;438, sep 1999. Reference for bacterial culture growth.
     </p>
     <p class="bibitem" ><span class="biblabel">
  [6]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XGriep:2006p719"></a>M&#x00A0;Griep, S&#x00A0;Whitney, M&#x00A0;Nelson, and H&#x00A0;Viljoen.   DNA polymerase chain
     reaction: A model of error frequencies and extension rates. <span 
class="cmti-12">AIChE journal</span>, 52(1),
     2006.
     </p>
     <p class="bibitem" ><span class="biblabel">
  [7]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XIhaka:1996p1018"></a>R&#x00A0;Ihaka and R&#x00A0;Gentleman.  R: A language for data analysis and graphics.
     <span 
class="cmti-12">Journal of computational and graphical statistics</span>, jan 1996.
     </p>
     <p class="bibitem" ><span class="biblabel">
  [8]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XLeipe:1999p1040"></a>D&#x00A0;D Leipe, L&#x00A0;Aravind, and E&#x00A0;V Koonin. Did DNA replication evolve twice
     independently? <span 
class="cmti-12">Nucleic acids research</span>, 27(17):3389&#8211;3401, sep 1999.
     </p>

     <p class="bibitem" ><span class="biblabel">
  [9]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XLevene:1934p869"></a>PA&#x00A0;Levene  and  ET&#x00A0;Stiller.   The  Synthesis  of  Ribose-5-phosphoric  Acid.
     <span 
class="cmti-12">Journal of Biological Chemistry</span>, 104(2):299&#8211;306, 1934.
     </p>
     <p class="bibitem" ><span class="biblabel">
 [10]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XANowak:2006p857"></a>Martin&#x00A0;A Nowak.  Evolutionary dynamics: exploring the equations of life.
     page 363, jan 2006.
     </p>
     <p class="bibitem" ><span class="biblabel">
 [11]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XPetruska:1988p721"></a>J&#x00A0;Petruska,  M&#x00A0;F  Goodman,  M&#x00A0;S  Boosalis,  L&#x00A0;C  Sowers,  C&#x00A0;Cheong,  and
     I&#x00A0;Tinoco.    Comparison  between  DNA  melting  thermodynamics  and  DNA
     polymerase fidelity.   <span 
class="cmti-12">Proceedings of the National Academy of Sciences of the</span>
     <span 
class="cmti-12">United States of America</span>, 85(17):6252&#8211;6256, sep 1988.  Reference for base-pair
     mismatch G.
     </p>
     <p class="bibitem" ><span class="biblabel">
 [12]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XSigel:1983p720"></a>H&#x00A0;Sigel  and  F&#x00A0;Hofstetter.   Metal-ion-promoted  dephosphorylation  of  the
     5&#8217;-triphosphates of uridine and thymidine, and a comparison with the reactivity
     in  the  corresponding  cytidine  and  adenosine  nucleotide  systems.   <span 
class="cmti-12">European</span>
     <span 
class="cmti-12">journal of biochemistry / FEBS</span>, 132(3):569&#8211;577, may 1983.
     </p>
     <p class="bibitem" ><span class="biblabel">
 [13]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XTempleton:1980p954"></a>A&#x00A0;R Templeton. The theory of speciation via the founder principle. <span 
class="cmti-12">Genetics</span>,
     94(4):1011&#8211;1038, apr 1980.

     </p>
     <p class="bibitem" ><span class="biblabel">
 [14]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XThieffry:1998p913"></a>D&#x00A0;Thieffry and S&#x00A0;Sarkar.  Forty years under the central dogma.  <span 
class="cmti-12">Trends in</span>
     <span 
class="cmti-12">biochemical sciences</span>, 23(8):312&#8211;316, aug 1998.
     </p>
     <p class="bibitem" ><span class="biblabel">
 [15]<span class="bibsp">&#x00A0;&#x00A0;&#x00A0;</span></span><a 
 id="XZannisHadjopoulos:1999p1929"></a>M&#x00A0;Zannis-Hadjopoulos and G&#x00A0;B Price. Eukaryotic DNA replication. <span 
class="cmti-12">Journal</span>
     <span 
class="cmti-12">of cellular biochemistry</span>, Suppl 32-33:1&#8211;14, jan 1999.
</p>
     </div>
        
</body></html> 



